--- ../src-base/minecraft/net/minecraft/entity/projectile/EntityLlamaSpit.java
+++ ../src-work/minecraft/net/minecraft/entity/projectile/EntityLlamaSpit.java
@@ -3,6 +3,9 @@
 import java.util.List;
 import java.util.UUID;
 import javax.annotation.Nullable;
+
+import carpet.CarpetSettings;
+import carpet.utils.RandomTools;
 import net.minecraft.block.material.Material;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.IProjectile;
@@ -155,9 +158,16 @@
         p_70186_1_ = p_70186_1_ / (double)f;
         p_70186_3_ = p_70186_3_ / (double)f;
         p_70186_5_ = p_70186_5_ / (double)f;
-        p_70186_1_ = p_70186_1_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double)p_70186_8_;
-        p_70186_3_ = p_70186_3_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double)p_70186_8_;
-        p_70186_5_ = p_70186_5_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double)p_70186_8_;
+        if (!CarpetSettings.extremeBehaviours) {
+            p_70186_1_ = p_70186_1_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double) p_70186_8_;
+            p_70186_3_ = p_70186_3_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double) p_70186_8_;
+            p_70186_5_ = p_70186_5_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double) p_70186_8_;
+        }
+        else {
+            p_70186_1_ = p_70186_1_ + RandomTools.nextNotGaussian(this.field_70146_Z) * 0.007499999832361937D * (double)p_70186_8_;
+            p_70186_3_ = p_70186_3_ + RandomTools.nextNotGaussian(this.field_70146_Z) * 0.007499999832361937D * (double)p_70186_8_;
+            p_70186_5_ = p_70186_5_ + RandomTools.nextNotGaussian(this.field_70146_Z) * 0.007499999832361937D * (double)p_70186_8_;
+        }
         p_70186_1_ = p_70186_1_ * (double)p_70186_7_;
         p_70186_3_ = p_70186_3_ * (double)p_70186_7_;
         p_70186_5_ = p_70186_5_ * (double)p_70186_7_;
