--- ../src-base/minecraft/net/minecraft/entity/projectile/EntityThrowable.java
+++ ../src-work/minecraft/net/minecraft/entity/projectile/EntityThrowable.java
@@ -3,6 +3,9 @@
 import java.util.List;
 import java.util.UUID;
 import javax.annotation.Nullable;
+
+import carpet.CarpetSettings;
+import carpet.utils.RandomTools;
 import net.minecraft.block.Block;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingBase;
@@ -21,6 +24,9 @@
 import net.minecraft.world.World;
 import net.minecraft.world.WorldServer;
 
+import carpet.logging.logHelpers.TrajectoryLogHelper;
+import carpet.logging.LoggerRegistry;
+
 public abstract class EntityThrowable extends Entity implements IProjectile
 {
     private int field_145788_c;
@@ -36,6 +42,8 @@
     public Entity field_184539_c;
     private int field_184540_av;
 
+    private TrajectoryLogHelper logHelper = null;
+
     public EntityThrowable(World p_i1776_1_)
     {
         super(p_i1776_1_);
@@ -43,6 +51,11 @@
         this.field_145786_d = -1;
         this.field_145787_e = -1;
         this.func_70105_a(0.25F, 0.25F);
+        //CM
+        if (LoggerRegistry.__projectiles)
+        {
+            logHelper = new TrajectoryLogHelper("projectiles");
+        }
     }
 
     public EntityThrowable(World p_i1778_1_, double p_i1778_2_, double p_i1778_4_, double p_i1778_6_)
@@ -82,9 +95,16 @@
         p_70186_1_ = p_70186_1_ / (double)f;
         p_70186_3_ = p_70186_3_ / (double)f;
         p_70186_5_ = p_70186_5_ / (double)f;
-        p_70186_1_ = p_70186_1_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double)p_70186_8_;
-        p_70186_3_ = p_70186_3_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double)p_70186_8_;
-        p_70186_5_ = p_70186_5_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double)p_70186_8_;
+        if (!CarpetSettings.extremeBehaviours) {
+            p_70186_1_ = p_70186_1_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double) p_70186_8_;
+            p_70186_3_ = p_70186_3_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double) p_70186_8_;
+            p_70186_5_ = p_70186_5_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double) p_70186_8_;
+        }
+        else {
+            p_70186_1_ = p_70186_1_ + RandomTools.nextNotGaussian(this.field_70146_Z) * 0.007499999832361937D * (double) p_70186_8_;
+            p_70186_3_ = p_70186_3_ + RandomTools.nextNotGaussian(this.field_70146_Z) * 0.007499999832361937D * (double) p_70186_8_;
+            p_70186_5_ = p_70186_5_ + RandomTools.nextNotGaussian(this.field_70146_Z) * 0.007499999832361937D * (double) p_70186_8_;
+        }
         p_70186_1_ = p_70186_1_ * (double)p_70186_7_;
         p_70186_3_ = p_70186_3_ * (double)p_70186_7_;
         p_70186_5_ = p_70186_5_ * (double)p_70186_7_;
@@ -268,9 +288,24 @@
             this.field_70181_x -= (double)f2;
         }
 
+        // ----- Carpet Start ----- //
+        if (LoggerRegistry.__projectiles && logHelper != null)
+        {
+            logHelper.onTick(field_70165_t, field_70163_u, field_70161_v, field_70159_w, field_70181_x, field_70179_y);
+        }
+        // ----- Carpet End ----- //
         this.func_70107_b(this.field_70165_t, this.field_70163_u, this.field_70161_v);
     }
 
+    @Override
+    public void func_70106_y() {
+        super.func_70106_y();
+        // ----- Carpet Start ----- //
+        if (LoggerRegistry.__projectiles && logHelper != null)
+            logHelper.onFinish();
+        // ----- Carpet End ----- //
+    }
+
     protected float func_70185_h()
     {
         return 0.03F;
@@ -331,6 +366,10 @@
     @Nullable
     public EntityLivingBase func_85052_h()
     {
+        // Fixes pearls disappearing when players relog similar to 1.15 CARPET-XCOM
+        if(CarpetSettings.fixedPearlBugs && pearlCheck()){
+            field_70192_c = null;
+        }
         if (this.field_70192_c == null && this.field_85053_h != null && !this.field_85053_h.isEmpty())
         {
             this.field_70192_c = this.field_70170_p.func_72924_a(this.field_85053_h);
@@ -355,4 +394,17 @@
 
         return this.field_70192_c;
     }
+
+    // Fixes pearls disappearing when players relog similar to 1.15 CARPET-XCOM
+    private boolean pearlCheck(){
+        if(field_85053_h == null){
+            if(field_70192_c == null) return false;
+            field_85053_h = this.field_70192_c.func_70005_c_();
+        }
+        Entity entity = null;
+        try {
+            entity = ((WorldServer) this.field_70170_p).func_175733_a(UUID.fromString(this.field_85053_h));
+        }catch(Exception e){return true;}
+        return !field_70170_p.field_73010_i.contains(entity);
+    }
 }
