--- ../src-base/minecraft/net/minecraft/world/end/DragonFightManager.java
+++ ../src-work/minecraft/net/minecraft/world/end/DragonFightManager.java
@@ -45,9 +45,15 @@
 import net.minecraft.world.gen.feature.WorldGenEndGateway;
 import net.minecraft.world.gen.feature.WorldGenEndPodium;
 import net.minecraft.world.gen.feature.WorldGenSpikes;
+
+import redstone.multimeter.common.TickTask;
+import redstone.multimeter.helper.WorldHelper;
+
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import carpet.CarpetSettings;
+
 public class DragonFightManager
 {
     private static final Logger field_186107_a = LogManager.getLogger();
@@ -149,6 +155,8 @@
 
     public void func_186105_b()
     {
+        WorldHelper.startTickTask(TickTask.DRAGON_FIGHT); // RSMM
+
         this.field_186109_c.func_186758_d(!this.field_186117_k);
 
         if (++this.field_186116_j >= 20)
@@ -244,6 +252,8 @@
                 }
             }
         }
+
+        WorldHelper.endTickTask(); // RSMM
     }
 
     protected void func_186095_a(DragonSpawnManager p_186095_1_)
@@ -437,6 +447,8 @@
             }
         }
 
+        // Fix for the end portal somehow spawning at y = -2 when spawning the first time in skyblock CARPET-XCOM
+        if(CarpetSettings.skyblock && field_186121_o.func_177956_o() <= 0) field_186121_o = new BlockPos(field_186121_o.func_177958_n(), 63, field_186121_o.func_177952_p());
         worldgenendpodium.func_180709_b(this.field_186110_d, new Random(), this.field_186121_o);
     }
 
