--- ../src-base/minecraft/net/minecraft/entity/player/EntityPlayerMP.java
+++ ../src-work/minecraft/net/minecraft/entity/player/EntityPlayerMP.java
@@ -114,6 +114,14 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import net.minecraft.entity.item.EntityMinecart;
+import net.minecraft.entity.passive.EntityLlama;
+
+import carpet.CarpetSettings;
+import carpet.helpers.EntityPlayerActionPack;
+import carpet.helpers.IPlayerSensitiveTileEntity;
+import carpet.logging.logHelpers.DamageReporter;
+
 public class EntityPlayerMP extends EntityPlayer implements IContainerListener
 {
     private static final Logger field_147102_bM = LogManager.getLogger();
@@ -153,6 +161,9 @@
     public int field_71138_i;
     public boolean field_71136_j;
 
+    //CM
+    public EntityPlayerActionPack actionPack;
+
     public EntityPlayerMP(MinecraftServer p_i45285_1_, WorldServer p_i45285_2_, GameProfile p_i45285_3_, PlayerInteractionManager p_i45285_4_)
     {
         super(p_i45285_2_, p_i45285_3_);
@@ -188,6 +199,9 @@
         {
             this.func_70107_b(this.field_70165_t, this.field_70163_u + 1.0D, this.field_70161_v);
         }
+
+        //CM
+        actionPack = new EntityPlayerActionPack(this);
     }
 
     public void func_70037_a(NBTTagCompound p_70037_1_)
@@ -313,6 +327,9 @@
 
     public void func_70071_h_()
     {
+        //CM
+        actionPack.onUpdate();
+        //CM end
         this.field_71134_c.func_73075_a();
         --this.field_147101_bU;
 
@@ -365,6 +382,13 @@
             }
         }
 
+        if (CarpetSettings.ridingPlayerUpdateFix) {
+            Entity riding = func_184208_bv();
+            if(riding != null && (riding instanceof EntityMinecart || riding instanceof EntityLlama)){
+                this.field_71133_b.func_184103_al().func_72358_d(this);
+            }
+        }
+
         CriteriaTriggers.field_193135_v.func_193182_a(this);
 
         if (this.field_193107_ct != null)
@@ -605,6 +629,7 @@
 
             if (!flag && this.field_147101_bU > 0 && p_70097_1_ != DamageSource.field_76380_i)
             {
+                DamageReporter.modify_damage(this, p_70097_1_, p_70097_2_, 0.0F, "respawn protection");
                 return false;
             }
             else
@@ -615,6 +640,7 @@
 
                     if (entity instanceof EntityPlayer && !this.func_96122_a((EntityPlayer)entity))
                     {
+                        DamageReporter.modify_damage(this, p_70097_1_, p_70097_2_, 0.0F, "PVP disabled");
                         return false;
                     }
 
@@ -624,6 +650,7 @@
 
                         if (entityarrow.field_70250_c instanceof EntityPlayer && !this.func_96122_a((EntityPlayer)entityarrow.field_70250_c))
                         {
+                            DamageReporter.modify_damage(this, p_70097_1_, p_70097_2_, 0.0F, "PVP disabled");
                             return false;
                         }
                     }
@@ -704,6 +731,8 @@
         if (p_147097_1_ != null)
         {
             SPacketUpdateTileEntity spacketupdatetileentity = p_147097_1_.func_189518_D_();
+            if (p_147097_1_ instanceof IPlayerSensitiveTileEntity)
+                spacketupdatetileentity = ((IPlayerSensitiveTileEntity) p_147097_1_).getUpdatePacketPlayerSensitive(this);
 
             if (spacketupdatetileentity != null)
             {
