--- ../src-base/minecraft/net/minecraft/entity/player/EntityPlayer.java
+++ ../src-work/minecraft/net/minecraft/entity/player/EntityPlayer.java
@@ -91,6 +91,13 @@
 import net.minecraft.world.World;
 import net.minecraft.world.WorldServer;
 
+import carpet.logging.logHelpers.KillLogHelper;
+import carpet.CarpetSettings;
+import carpet.logging.logHelpers.DamageReporter;
+import carpet.helpers.PortalHelper;
+import carpet.logging.LoggerRegistry;
+import carpet.helpers.StatHelper;
+
 public abstract class EntityPlayer extends EntityLivingBase
 {
     private static final DataParameter<Float> field_184829_a = EntityDataManager.<Float>func_187226_a(EntityPlayer.class, DataSerializers.field_187193_c);
@@ -364,6 +371,15 @@
 
     public int func_82145_z()
     {
+        //CM
+        if (CarpetSettings.portalCreativeDelay)
+        {
+            if ( PortalHelper.player_holds_obsidian(this))
+            {
+                return 72000;
+            }
+            return 80;
+        }
         return this.field_71075_bZ.field_75102_a ? 1 : 80;
     }
 
@@ -530,7 +546,9 @@
         this.func_192028_j(this.func_192023_dk());
         this.func_192028_j(this.func_192025_dl());
 
-        if (!this.field_70170_p.field_72995_K && (this.field_70143_R > 0.5F || this.func_70090_H() || this.func_184218_aH()) || this.field_71075_bZ.field_75100_b)
+        //CM conditions
+        boolean parrots_will_drop = !CarpetSettings.persistentParrots || this.field_71075_bZ.field_75102_a;
+        if (!this.field_70170_p.field_72995_K && ((parrots_will_drop && this.field_70143_R > 0.5F) || this.func_70090_H() || (parrots_will_drop && this.func_184218_aH()) ) || this.field_71075_bZ.field_75100_b )
         {
             this.func_192030_dh();
         }
@@ -692,7 +710,7 @@
             {
                 if (!itemstack.func_190926_b())
                 {
-                    this.func_71064_a(StatList.func_188058_e(itemstack.func_77973_b()), p_146097_1_.func_190916_E());
+                    this.func_71064_a(StatHelper.getDroppedObjectStats(itemstack.func_77973_b(), itemstack.func_77960_j()), p_146097_1_.func_190916_E());
                 }
 
                 this.func_71029_a(StatList.field_75952_v);
@@ -908,10 +926,23 @@
                     this.func_70999_a(true, true, false);
                 }
 
-                this.func_192030_dh();
+                //this.spawnShoulderEntities();
+                //CM optionally dismount parrots
+                if (CarpetSettings.persistentParrots && !this.func_70093_af())
+                {
+                    if (this.field_70146_Z.nextFloat() < ((p_70097_2_)/15.0) )
+                    {
+                        this.dismount_left();
+                    }
+                    if (this.field_70146_Z.nextFloat() < ((p_70097_2_)/15.0) )
+                    {
+                        this.dismount_right();
+                    }
+                }
 
                 if (p_70097_1_.func_76350_n())
                 {
+                    float prevamount = p_70097_2_; //CM
                     if (this.field_70170_p.func_175659_aa() == EnumDifficulty.PEACEFUL)
                     {
                         p_70097_2_ = 0.0F;
@@ -926,6 +957,7 @@
                     {
                         p_70097_2_ = p_70097_2_ * 3.0F / 2.0F;
                     }
+                    DamageReporter.modify_damage(this, p_70097_1_, prevamount, p_70097_2_, "difficulty"); //CM
                 }
 
                 return p_70097_2_ == 0.0F ? false : super.func_70097_a(p_70097_1_, p_70097_2_);
@@ -1008,16 +1040,22 @@
     {
         if (!this.func_180431_b(p_70665_1_))
         {
+            float previous_amount = p_70665_2_; //CM
             p_70665_2_ = this.func_70655_b(p_70665_1_, p_70665_2_);
+            DamageReporter.modify_damage(this, p_70665_1_, previous_amount, p_70665_2_, //CM
+                    String.format("armour %.1f and toughness %.1f", (float)this.func_70658_aO(), (float)this.func_110148_a(SharedMonsterAttributes.field_189429_h).func_111126_e())); //CM
+
             p_70665_2_ = this.func_70672_c(p_70665_1_, p_70665_2_);
             float f = p_70665_2_;
             p_70665_2_ = Math.max(p_70665_2_ - this.func_110139_bj(), 0.0F);
             this.func_110149_m(this.func_110139_bj() - (f - p_70665_2_));
+            DamageReporter.modify_damage(this, p_70665_1_, p_70665_2_, f, "Absorbtion"); //CM
 
             if (p_70665_2_ != 0.0F)
             {
                 this.func_71020_j(p_70665_1_.func_76345_d());
                 float f1 = this.func_110143_aJ();
+                DamageReporter.register_final_damage(this, p_70665_1_, p_70665_2_); //CM
                 this.func_70606_j(this.func_110143_aJ() - p_70665_2_);
                 this.func_110142_aN().func_94547_a(p_70665_1_, f1, p_70665_2_);
 
@@ -1027,6 +1065,12 @@
                 }
             }
         }
+        //CM 
+        else
+        {
+            DamageReporter.modify_damage(this, p_70665_1_, p_70665_2_, 0.0f, "invulnerability to the damage source");
+        }
+        //CM end
     }
 
     public void func_175141_a(TileEntitySign p_175141_1_)
@@ -1227,19 +1271,33 @@
                         {
                             float f3 = 1.0F + EnchantmentHelper.func_191527_a(this) * f;
 
+                            //CM
+                            int mobs_smashed = 1;
+
                             for (EntityLivingBase entitylivingbase : this.field_70170_p.func_72872_a(EntityLivingBase.class, p_71059_1_.func_174813_aQ().func_72314_b(1.0D, 0.25D, 1.0D)))
                             {
                                 if (entitylivingbase != this && entitylivingbase != p_71059_1_ && !this.func_184191_r(entitylivingbase) && this.func_70068_e(entitylivingbase) < 9.0D)
                                 {
                                     entitylivingbase.func_70653_a(this, 0.4F, (double)MathHelper.func_76126_a(this.field_70177_z * 0.017453292F), (double)(-MathHelper.func_76134_b(this.field_70177_z * 0.017453292F)));
                                     entitylivingbase.func_70097_a(DamageSource.func_76365_a(this), f3);
+                                    //CM
+                                    mobs_smashed ++;
                                 }
                             }
 
+                            if (LoggerRegistry.__kills)
+                                KillLogHelper.onSweep(this, mobs_smashed);
+
                             this.field_70170_p.func_184148_a((EntityPlayer)null, this.field_70165_t, this.field_70163_u, this.field_70161_v, SoundEvents.field_187730_dW, this.func_184176_by(), 1.0F, 1.0F);
                             this.func_184810_cG();
                         }
+                        else
+                        {
+                            if (LoggerRegistry.__kills)
+                                KillLogHelper.onNonSweepAttack(this);
+                        }
 
+
                         if (p_71059_1_ instanceof EntityPlayerMP && p_71059_1_.field_70133_I)
                         {
                             ((EntityPlayerMP)p_71059_1_).field_71135_a.func_147359_a(new SPacketEntityVelocity(p_71059_1_));
@@ -1333,6 +1391,13 @@
                     }
                 }
             }
+            //CM
+            else
+            {
+                if (LoggerRegistry.__kills)
+                    KillLogHelper.onDudHit(this);
+            }
+            //CM END
         }
     }
 
@@ -2044,6 +2109,20 @@
         this.func_192031_i(new NBTTagCompound());
     }
 
+    //CM dismount left parrot
+    protected void dismount_left()
+    {
+        this.func_192026_k(this.func_192023_dk());
+        this.func_192029_h(new NBTTagCompound());
+    }
+
+    //dismount right parrot
+    protected void dismount_right()
+    {
+        this.func_192026_k(this.func_192025_dl());
+        this.func_192031_i(new NBTTagCompound());
+    }
+
     private void func_192026_k(@Nullable NBTTagCompound p_192026_1_)
     {
         if (!this.field_70170_p.field_72995_K && !p_192026_1_.func_82582_d())
@@ -2307,6 +2386,13 @@
     {
         return this.field_71075_bZ.field_75098_d && this.func_70003_b(2, "");
     }
+    
+    // CM: player collision
+    @Override
+    public boolean func_70104_M()
+    {
+        return !CarpetSettings.disablePlayerCollision;
+    }
 
     public static enum EnumChatVisibility
     {
@@ -2333,11 +2419,11 @@
         }
     }
 
-    static class SleepEnemyPredicate implements Predicate<EntityMob>
+    public static class SleepEnemyPredicate implements Predicate<EntityMob>
         {
             private final EntityPlayer field_192387_a;
 
-            private SleepEnemyPredicate(EntityPlayer p_i47461_1_)
+            public SleepEnemyPredicate(EntityPlayer p_i47461_1_)
             {
                 this.field_192387_a = p_i47461_1_;
             }
