--- ../src-base/minecraft/net/minecraft/world/biome/Biome.java
+++ ../src-work/minecraft/net/minecraft/world/biome/Biome.java
@@ -520,6 +520,12 @@
                 this.field_76299_d = p_i1970_4_;
             }
 
+            //carpet mod
+            public int getWeight()
+            {
+                return this.field_76292_a;
+            }
+
             public String toString()
             {
                 return this.field_76300_b.getSimpleName() + "*(" + this.field_76301_c + "-" + this.field_76299_d + "):" + this.field_76292_a;
