--- ../src-base/minecraft/net/minecraft/entity/passive/EntityLlama.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityLlama.java
@@ -47,6 +47,10 @@
 import net.minecraft.world.World;
 import net.minecraft.world.storage.loot.LootTableList;
 
+import carpet.CarpetSettings;
+
+ import carpet.CarpetSettings;
+
 public class EntityLlama extends AbstractChestHorse implements IRangedAttackMob
 {
     private static final DataParameter<Integer> field_190720_bG = EntityDataManager.<Integer>func_187226_a(EntityLlama.class, DataSerializers.field_187192_b);
@@ -192,7 +196,7 @@
             j = 6;
             f = 10.0F;
 
-            if (this.func_110248_bS() && this.func_70874_b() == 0)
+            if (this.func_110248_bS() && this.func_70874_b() == 0 && !(CarpetSettings.llamaOverfeedingFix && this.func_70880_s() ) )
             {
                 flag = true;
                 this.func_146082_f(p_190678_1_);
@@ -492,6 +496,19 @@
         }
     }
 
+    /*
+     * Breeding items for llamas: CARPET-XCOM
+     */
+    @Override
+    protected boolean isValidBreedingItem(ItemStack itemstack){
+    	Item item = itemstack.func_77973_b();
+
+    	if(item == Items.field_151015_O) return false;
+    	else if(item == Item.func_150898_a(Blocks.field_150407_cf)) return false;
+
+    	return true;
+    }
+
     public boolean func_190684_dE()
     {
         return false;
