--- ../src-base/minecraft/net/minecraft/world/gen/structure/template/Template.java
+++ ../src-work/minecraft/net/minecraft/world/gen/structure/template/Template.java
@@ -36,6 +36,10 @@
 import net.minecraft.world.World;
 import net.minecraft.world.gen.structure.StructureBoundingBox;
 
+import carpet.CarpetSettings;
+
+ import carpet.CarpetSettings;
+
 public class Template
 {
     private final List<Template.BlockInfo> field_186270_a = Lists.<Template.BlockInfo>newArrayList();
@@ -233,11 +237,11 @@
                                     ((IInventory)tileentity).func_174888_l();
                                 }
 
-                                p_189960_1_.func_180501_a(blockpos, Blocks.field_180401_cv.func_176223_P(), 4);
+                                p_189960_1_.func_180501_a(blockpos, Blocks.field_180401_cv.func_176223_P(), 4 | (CarpetSettings.fillUpdates?0:128)); //CM
                             }
                         }
 
-                        if (p_189960_1_.func_180501_a(blockpos, iblockstate1, p_189960_5_) && template$blockinfo1.field_186244_c != null)
+                        if (p_189960_1_.func_180501_a(blockpos, iblockstate1, p_189960_5_ | (CarpetSettings.fillUpdates?0:128)) && template$blockinfo1.field_186244_c != null) //CM
                         {
                             TileEntity tileentity2 = p_189960_1_.func_175625_s(blockpos);
 
@@ -263,8 +267,13 @@
 
                     if (structureboundingbox == null || structureboundingbox.func_175898_b(blockpos1))
                     {
+                        //CARPET MOD ADDED CONDITION
+                        if (CarpetSettings.fillUpdates)
+                        {
                         p_189960_1_.func_175722_b(blockpos1, template$blockinfo2.field_186243_b.func_177230_c(), false);
-
+                        }
+                        //CM end
+                        
                         if (template$blockinfo2.field_186244_c != null)
                         {
                             TileEntity tileentity1 = p_189960_1_.func_175625_s(blockpos1);
