--- ../src-base/minecraft/net/minecraft/tileentity/TileEntity.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntity.java
@@ -1,6 +1,8 @@
 package net.minecraft.tileentity;
 
 import javax.annotation.Nullable;
+
+import carpet.helpers.TileEntityCraftingTable;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockJukebox;
 import net.minecraft.block.state.IBlockState;
@@ -307,5 +309,6 @@
         func_190560_a("command_block", TileEntityCommandBlock.class);
         func_190560_a("shulker_box", TileEntityShulkerBox.class);
         func_190560_a("bed", TileEntityBed.class);
+        func_190560_a("autocrafting", TileEntityCraftingTable.class);
     }
 }
