--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityPiston.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityPiston.java
@@ -20,6 +20,8 @@
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
 
+import carpet.CarpetSettings;
+
 public class TileEntityPiston extends TileEntity implements ITickable
 {
     private IBlockState field_174932_a;
@@ -166,6 +168,11 @@
                             {
                                 this.func_190605_a(entity, enumfacing, d0);
                             }
+
+                            if(CarpetSettings.unloadedEntityFix)
+                            {
+                                field_145850_b.func_72866_a(entity, false);
+                            }
                         }
                     }
                 }
