--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityFurnace.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityFurnace.java
@@ -29,7 +29,10 @@
 import net.minecraft.util.datafix.walkers.ItemStackDataLists;
 import net.minecraft.util.math.MathHelper;
 
-public class TileEntityFurnace extends TileEntityLockable implements ITickable, ISidedInventory
+import carpet.CarpetSettings;
+import carpet.helpers.TileEntityOptimizer.ILazyTileEntity;
+
+public class TileEntityFurnace extends TileEntityLockable implements ITickable, ISidedInventory, ILazyTileEntity
 {
     private static final int[] field_145962_k = new int[] {0};
     private static final int[] field_145959_l = new int[] {2, 1};
@@ -41,6 +44,9 @@
     private int field_174905_l;
     private String field_145958_o;
 
+
+    //CM
+    public String cm_name() { return "Furnace"; }
     public int func_70302_i_()
     {
         return this.field_145957_n.size();
@@ -157,6 +163,10 @@
 
     public void func_73660_a()
     {
+        // CARPET-optimizedTileEntities: Skip the update if sleeping
+        if(CarpetSettings.optimizedTileEntities && this.isSleeping)
+            return;
+
         boolean flag = this.func_145950_i();
         boolean flag1 = false;
 
@@ -221,6 +231,10 @@
                 flag1 = true;
                 BlockFurnace.func_176446_a(this.func_145950_i(), this.field_145850_b, this.field_174879_c);
             }
+
+            // CARPET-optimizedTileEntities: Fall asleep if not burning
+            if(!this.func_145950_i())
+                this.isSleeping = true;
         }
 
         if (flag1)
@@ -521,4 +535,14 @@
     {
         this.field_145957_n.clear();
     }
+
+    // CARPET-optimizedTileEntities: Whether the tile entity is asleep or not.
+    // False by default so tile entities wake up upon chunk loading
+    private boolean isSleeping = false;
+
+    // CARPET-optimizedTileEntities
+    @Override
+    public void wakeUp(){
+        this.isSleeping = false;
+    }
 }
