--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityBrewingStand.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityBrewingStand.java
@@ -23,7 +23,10 @@
 import net.minecraft.util.datafix.walkers.ItemStackDataLists;
 import net.minecraft.util.math.BlockPos;
 
-public class TileEntityBrewingStand extends TileEntityLockable implements ITickable, ISidedInventory
+import carpet.CarpetSettings;
+import carpet.helpers.TileEntityOptimizer.ILazyTileEntity;
+
+public class TileEntityBrewingStand extends TileEntityLockable implements ITickable, ISidedInventory, ILazyTileEntity
 {
     private static final int[] field_145941_a = new int[] {3};
     private static final int[] field_184277_f = new int[] {0, 1, 2, 3};
@@ -70,6 +73,10 @@
 
     public void func_73660_a()
     {
+        // CARPET-optimizedTileEntities: Skip the update if sleeping
+        if(CarpetSettings.optimizedTileEntities && this.isSleeping)
+            return;
+
         ItemStack itemstack = this.field_145945_j.get(4);
 
         if (this.field_184278_m <= 0 && itemstack.func_77973_b() == Items.field_151065_br)
@@ -134,6 +141,10 @@
                 this.field_145850_b.func_180501_a(this.field_174879_c, iblockstate, 2);
             }
         }
+
+        // CARPET-optimizedTileEntities: Fall asleep if not brewing
+        if(this.field_145946_k == 0)
+            this.isSleeping = true;
     }
 
     public boolean[] func_174902_m()
@@ -386,4 +397,12 @@
     {
         this.field_145945_j.clear();
     }
+
+    // CARPET-optimizedTileEntities: Whether the tile entity is asleep or not.
+    // False by default so tile entities wake up upon chunk loading
+    private boolean isSleeping = false;
+
+    // CARPET-optimizedTileEntities
+    @Override
+    public void wakeUp() { this.isSleeping = false; }
 }
