--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityChest.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityChest.java
@@ -203,7 +203,7 @@
         }
         else
         {
-            Block block = this.field_145850_b.func_180495_p(p_174912_1_).func_177230_c();
+            Block block = this.field_145850_b.getBlockState(p_174912_1_, "Chest loading").func_177230_c();
             return block instanceof BlockChest && ((BlockChest)block).field_149956_a == this.func_145980_j();
         }
     }
