--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityComparator.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityComparator.java
@@ -5,6 +5,10 @@
 public class TileEntityComparator extends TileEntity
 {
     private int field_145997_a;
+    // CM: instant comparator logger, stored in world time modulo 3.
+    // This is to allow for further tile tick scheduling in the same tick before the tile tick is processed
+    public int[] scheduledOutputSignal = new int[3];
+    public boolean[] buggy = new boolean[3];
 
     public NBTTagCompound func_189515_b(NBTTagCompound p_189515_1_)
     {
