--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityBeacon.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityBeacon.java
@@ -32,6 +32,8 @@
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
 
+import carpet.CarpetSettings;
+
 public class TileEntityBeacon extends TileEntityLockable implements ITickable, ISidedInventory
 {
     public static final Potion[][] field_146009_a = new Potion[][] {{MobEffects.field_76424_c, MobEffects.field_76422_e}, {MobEffects.field_76429_m, MobEffects.field_76430_j}, {MobEffects.field_76420_g}, {MobEffects.field_76428_l}};
@@ -111,6 +113,10 @@
         boolean flag = true;
         BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
 
+        // CARPET-optimizedTileEntities
+        if(!(CarpetSettings.optimizedTileEntities && !this.field_145850_b.field_72995_K
+                && this.field_145850_b.field_73011_w.func_191066_m() && this.field_145850_b.func_175678_i(field_174879_c)))
+        {
         for (int i1 = j + 1; i1 < 256; ++i1)
         {
             IBlockState iblockstate = this.field_145850_b.func_180495_p(blockpos$mutableblockpos.func_181079_c(i, i1, k));
@@ -155,6 +161,7 @@
 
             flag = false;
         }
+        }
 
         if (this.field_146015_k)
         {
