--- ../src-base/minecraft/net/minecraft/entity/Entity.java
+++ ../src-work/minecraft/net/minecraft/entity/Entity.java
@@ -81,6 +81,10 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import carpet.CarpetSettings;
+import carpet.helpers.BabyGrowingUp;
+import carpet.patches.EntityPlayerMPFake;
+
 public abstract class Entity implements ICommandSender
 {
     private static final Logger field_184243_a = LogManager.getLogger();
@@ -278,6 +282,14 @@
     {
         if (p_70105_1_ != this.field_70130_N || p_70105_2_ != this.field_70131_O)
         {
+            /*
+             * Redirected code for changing size of mobs CARPET-XCOM
+             */
+            if(CarpetSettings.growingUpWallJumpFix){
+                BabyGrowingUp.carpetSetSize(this, p_70105_1_, p_70105_2_);
+                return;
+            }
+
             float f = this.field_70130_N;
             this.field_70130_N = p_70105_1_;
             this.field_70131_O = p_70105_2_;
@@ -654,12 +666,28 @@
                 }
             }
 
-            List<AxisAlignedBB> list1 = this.field_70170_p.func_184144_a(this, this.func_174813_aQ().func_72321_a(p_70091_2_, p_70091_4_, p_70091_6_));
+            // Entity move optimization for faster entities CARPET-MASA
+            List<AxisAlignedBB> list1 = null;
+            boolean optimize = false;
+            if (CarpetSettings.fastMovingEntityOptimization &&
+                    (p_70091_2_ > 4 || p_70091_2_ < -4 ||
+                     p_70091_4_ > 4 || p_70091_4_ < -4 ||
+                     p_70091_6_ > 4 || p_70091_6_ < -4)) {
+
+                optimize = true;
+            }
+
+            if (!optimize) { // CARPET-MASA
+                list1 = this.field_70170_p.func_184144_a(this, this.func_174813_aQ().func_72321_a(p_70091_2_, p_70091_4_, p_70091_6_));
+            }
             AxisAlignedBB axisalignedbb = this.func_174813_aQ();
 
             if (p_70091_4_ != 0.0D)
             {
                 int k = 0;
+                if (optimize) { // CARPET-MASA
+                    list1 = this.field_70170_p.func_184144_a(this, this.func_174813_aQ().func_72321_a(0, p_70091_4_, 0));
+                }
 
                 for (int l = list1.size(); k < l; ++k)
                 {
@@ -672,6 +700,9 @@
             if (p_70091_2_ != 0.0D)
             {
                 int j5 = 0;
+                if (optimize) { // CARPET-MASA
+                    list1 = this.field_70170_p.func_184144_a(this, this.func_174813_aQ().func_72321_a(p_70091_2_, 0, 0));
+                }
 
                 for (int l5 = list1.size(); j5 < l5; ++j5)
                 {
@@ -687,6 +718,9 @@
             if (p_70091_6_ != 0.0D)
             {
                 int k5 = 0;
+                if (optimize) { // CARPET-MASA
+                    list1 = this.field_70170_p.func_184144_a(this, this.func_174813_aQ().func_72321_a(0, 0, p_70091_6_));
+                }
 
                 for (int i6 = list1.size(); k5 < i6; ++k5)
                 {
@@ -1533,6 +1567,9 @@
 
         if (!this.field_70128_L && s != null && !this.func_184218_aH())
         {
+            // Fix for fixing duplication caused by riding entitys into unloaded chunks CARPET-XCOM
+            if(CarpetSettings.duplicationFixRidingEntitys && hasPlayerPassenger(this)) return false;
+
             p_70039_1_.func_74778_a("id", s);
             this.func_189511_e(p_70039_1_);
             return true;
@@ -1543,6 +1580,19 @@
         }
     }
 
+    // Method for fixing duplication caused by riding entitys into unloaded chunks CARPET-XCOM
+    private static boolean hasPlayerPassenger(Entity entity) {
+        for (Entity passenger : entity.field_184244_h) {
+            if (passenger instanceof EntityPlayer && !(passenger instanceof EntityPlayerMPFake)) {
+                return true;
+            }
+            if (hasPlayerPassenger(passenger)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
     public static void func_190533_a(DataFixer p_190533_0_)
     {
         p_190533_0_.func_188258_a(FixTypes.ENTITY, new IDataWalker()
