--- ../src-base/minecraft/net/minecraft/entity/Entity.java
+++ ../src-work/minecraft/net/minecraft/entity/Entity.java
@@ -81,6 +81,10 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import carpet.CarpetSettings;
+import carpet.helpers.BabyGrowingUp;
+import carpet.helpers.BlockRotator;
+
 public abstract class Entity implements ICommandSender
 {
     private static final Logger field_184243_a = LogManager.getLogger();
@@ -136,7 +140,7 @@
     private int field_190534_ay;
     protected boolean field_70171_ac;
     public int field_70172_ad;
-    protected boolean field_70148_d;
+    public boolean field_70148_d;
     protected boolean field_70178_ae;
     protected EntityDataManager field_70180_af;
     protected static final DataParameter<Byte> field_184240_ax = EntityDataManager.<Byte>func_187226_a(Entity.class, DataSerializers.field_187191_a);
@@ -168,6 +172,9 @@
     private final double[] field_191505_aI;
     private long field_191506_aJ;
 
+        //CM
+    public String cm_name() { return "Other Entity"; }
+
     public Entity(World p_i1582_1_)
     {
         this.field_145783_c = field_70152_a++;
@@ -278,6 +285,14 @@
     {
         if (p_70105_1_ != this.field_70130_N || p_70105_2_ != this.field_70131_O)
         {
+            /*
+             * Redirected code for changing size of mobs CARPET-XCOM
+             */
+            if(CarpetSettings.getBool("growingUpWallJump") ){
+                BabyGrowingUp.carpetSetSize(this, p_70105_1_, p_70105_2_);
+                return;
+            }
+
             float f = this.field_70130_N;
             this.field_70130_N = p_70105_1_;
             this.field_70131_O = p_70105_2_;
@@ -299,7 +314,8 @@
         }
     }
 
-    protected void func_70101_b(float p_70101_1_, float p_70101_2_)
+    //CM changed from protected to set in player action pack
+    public void func_70101_b(float p_70101_1_, float p_70101_2_)
     {
         this.field_70177_z = p_70101_1_ % 360.0F;
         this.field_70125_A = p_70101_2_ % 360.0F;
@@ -654,12 +670,28 @@
                 }
             }
 
-            List<AxisAlignedBB> list1 = this.field_70170_p.func_184144_a(this, this.func_174813_aQ().func_72321_a(p_70091_2_, p_70091_4_, p_70091_6_));
+            // Entity move optimization for faster entities CARPET-MASA
+            List<AxisAlignedBB> list1 = null;
+            boolean optimize = false;
+            if (CarpetSettings.fastMovingEntityOptimization &&
+                    (p_70091_2_ > 4 || p_70091_2_ < -4 ||
+                     p_70091_4_ > 4 || p_70091_4_ < -4 ||
+                     p_70091_6_ > 4 || p_70091_6_ < -4)) {
+
+                optimize = true;
+            }
+
+            if (!optimize) { // CARPET-MASA
+                list1 = this.field_70170_p.func_184144_a(this, this.func_174813_aQ().func_72321_a(p_70091_2_, p_70091_4_, p_70091_6_));
+            }
             AxisAlignedBB axisalignedbb = this.func_174813_aQ();
 
             if (p_70091_4_ != 0.0D)
             {
                 int k = 0;
+                if (optimize) { // CARPET-MASA
+                    list1 = this.field_70170_p.func_184144_a(this, this.func_174813_aQ().func_72321_a(0, p_70091_4_, 0));
+                }
 
                 for (int l = list1.size(); k < l; ++k)
                 {
@@ -672,6 +704,9 @@
             if (p_70091_2_ != 0.0D)
             {
                 int j5 = 0;
+                if (optimize) { // CARPET-MASA
+                    list1 = this.field_70170_p.func_184144_a(this, this.func_174813_aQ().func_72321_a(p_70091_2_, 0, 0));
+                }
 
                 for (int l5 = list1.size(); j5 < l5; ++j5)
                 {
@@ -687,6 +722,9 @@
             if (p_70091_6_ != 0.0D)
             {
                 int k5 = 0;
+                if (optimize) { // CARPET-MASA
+                    list1 = this.field_70170_p.func_184144_a(this, this.func_174813_aQ().func_72321_a(0, 0, p_70091_6_));
+                }
 
                 for (int i6 = list1.size(); k5 < i6; ++k5)
                 {
@@ -1908,7 +1946,12 @@
 
     public double func_70042_X()
     {
+        if (CarpetSettings.getBool("rideableGhasts") && this.field_70131_O > 2.0F)
+        {
+            return (double)this.field_70131_O - 0.2D;
+        }
         return (double)this.field_70131_O * 0.75D;
+
     }
 
     public boolean func_184220_m(Entity p_184220_1_)
@@ -2578,6 +2621,10 @@
 
     public EnumFacing func_174811_aO()
     {
+        //CM
+        if (BlockRotator.flippinEligibility(this))
+            return EnumFacing.func_176731_b(MathHelper.func_76128_c((double)(this.field_70177_z * 4.0F / 360.0F) + 0.5D) & 3).func_176734_d();
+
         return EnumFacing.func_176731_b(MathHelper.func_76128_c((double)(this.field_70177_z * 4.0F / 360.0F) + 0.5D) & 3);
     }
 
@@ -2886,4 +2933,14 @@
     {
         return 1;
     }
+
+    //CM
+    public int getFire()
+    {
+        return this.field_190534_ay;
+    }
+    
+    public void postLoad()
+    {
+    }
 }
