--- ../src-base/minecraft/net/minecraft/entity/EntityLivingBase.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityLivingBase.java
@@ -75,6 +75,9 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import carpet.CarpetSettings;
+import carpet.logging.logHelpers.DamageReporter;
+
 public abstract class EntityLivingBase extends Entity
 {
     private static final Logger field_190632_a = LogManager.getLogger();
@@ -151,6 +154,10 @@
     private DamageSource field_189750_bF;
     private long field_189751_bG;
 
+    //CM
+    public long birthTick;
+    public String cm_name() { return "Other Living Entity"; }
+
     public void func_174812_G()
     {
         this.func_70097_a(DamageSource.field_76380_i, Float.MAX_VALUE);
@@ -168,6 +175,8 @@
         this.field_70177_z = (float)(Math.random() * (Math.PI * 2D));
         this.field_70759_as = this.field_70177_z;
         this.field_70138_W = 0.6F;
+        //CM
+        this.birthTick = p_i1594_1_.func_82737_E();
     }
 
     protected void func_70088_a()
@@ -401,6 +410,7 @@
 
     protected boolean func_146066_aG()
     {
+        if(CarpetSettings.duplicationFixPortalEntitys && field_70128_L) return false; // Fixes duplication when entity's are set to be removed and can still drop items.
         return !this.func_70631_g_();
     }
 
@@ -741,8 +751,16 @@
             }
             else
             {
-                potioneffect.func_76452_a(p_70690_1_);
-                this.func_70695_b(potioneffect, true);
+                PotionEffect newEffect = potioneffect.combine(p_70690_1_);
+                if (newEffect != potioneffect) {
+                    // carpet
+                    this.field_70713_bf.put(newEffect.func_188419_a(), newEffect);
+                    this.func_70688_c(potioneffect);
+                    this.func_70670_a(newEffect);
+                } else {
+                    // vanilla
+                    this.func_70695_b(newEffect, true);
+                }
             }
         }
     }
@@ -813,6 +831,11 @@
         {
             p_70688_1_.func_188419_a().func_111187_a(this, this.func_110140_aT(), p_70688_1_.func_76458_c());
         }
+        // CM
+        if (CarpetSettings.effectsFix && p_70688_1_.previous != null) {
+            func_70690_d(p_70688_1_.previous);
+        }
+        // CM END
     }
 
     public void func_70691_i(float p_70691_1_)
@@ -847,14 +870,20 @@
         }
         else
         {
-            this.field_70708_bq = 0;
+            DamageReporter.register_damage(this, p_70097_1_, p_70097_2_); //CM
+            if (!CarpetSettings.potionsDespawnFix) // CM run conditionally with vanilla behaviour
+            {
+                this.field_70708_bq = 0;
+            }
 
             if (this.func_110143_aJ() <= 0.0F)
             {
+                DamageReporter.modify_damage(this, p_70097_1_, p_70097_2_, 0.0f, "Already dead and can't take more damage"); //CM
                 return false;
             }
             else if (p_70097_1_.func_76347_k() && this.func_70644_a(MobEffects.field_76426_n))
             {
+                DamageReporter.modify_damage(this, p_70097_1_, p_70097_2_, 0.0f, "Resistance to fire"); //CM
                 return false;
             }
             else
@@ -864,6 +893,7 @@
                 if ((p_70097_1_ == DamageSource.field_82728_o || p_70097_1_ == DamageSource.field_82729_p) && !this.func_184582_a(EntityEquipmentSlot.HEAD).func_190926_b())
                 {
                     this.func_184582_a(EntityEquipmentSlot.HEAD).func_77972_a((int)(p_70097_2_ * 4.0F + this.field_70146_Z.nextFloat() * p_70097_2_ * 2.0F), this);
+                    DamageReporter.modify_damage(this, p_70097_1_, p_70097_2_, p_70097_2_*0.75f, "wearing a helmet"); //CM
                     p_70097_2_ *= 0.75F;
                 }
 
@@ -872,6 +902,7 @@
                 if (p_70097_2_ > 0.0F && this.func_184583_d(p_70097_1_))
                 {
                     this.func_184590_k(p_70097_2_);
+                    DamageReporter.modify_damage(this, p_70097_1_, p_70097_2_, 0.0f, "using a shield"); //CM
                     p_70097_2_ = 0.0F;
 
                     if (!p_70097_1_.func_76352_a())
@@ -894,9 +925,10 @@
                 {
                     if (p_70097_2_ <= this.field_110153_bc)
                     {
+                        DamageReporter.modify_damage(this, p_70097_1_, p_70097_2_, 0.0f, "Recently hit"); //CM
                         return false;
                     }
-
+                    DamageReporter.modify_damage(this, p_70097_1_, p_70097_2_, p_70097_2_ - this.field_110153_bc, "Recently hit"); //CM
                     this.func_70665_d(p_70097_1_, p_70097_2_ - this.field_110153_bc);
                     this.field_110153_bc = p_70097_2_;
                     flag1 = false;
@@ -1009,6 +1041,11 @@
                 {
                     this.func_184581_c(p_70097_1_);
                 }
+                if (CarpetSettings.potionsDespawnFix)
+                    // CM reset entity age is connected to making a hurt noise
+                {
+                    this.field_70708_bq = 0;
+                }
 
                 boolean flag2 = !flag || p_70097_2_ > 0.0F;
 
@@ -1345,6 +1382,7 @@
                 int i = (this.func_70660_b(MobEffects.field_76429_m).func_76458_c() + 1) * 5;
                 int j = 25 - i;
                 float f = p_70672_2_ * (float)j;
+                DamageReporter.modify_damage(this, p_70672_1_, p_70672_2_, f / 25.0F, "Resistance status effect"); //CM
                 p_70672_2_ = f / 25.0F;
             }
 
@@ -1358,7 +1396,10 @@
 
                 if (k > 0)
                 {
+                    float previous_damage = p_70672_2_;
                     p_70672_2_ = CombatRules.func_188401_b(p_70672_2_, (float)k);
+                    DamageReporter.modify_damage(this, p_70672_1_, previous_damage, p_70672_2_,
+                            String.format("enchantments (%.1f total points)", (float)k));
                 }
 
                 return p_70672_2_;
@@ -1370,15 +1411,21 @@
     {
         if (!this.func_180431_b(p_70665_1_))
         {
+            float previous_amount = p_70665_2_; //CM
             p_70665_2_ = this.func_70655_b(p_70665_1_, p_70665_2_);
+            DamageReporter.modify_damage(this, p_70665_1_, previous_amount, p_70665_2_, //CM
+                    String.format("Armour %.1f, Toughness %.1f", (float)this.func_70658_aO(), (float)this.func_110148_a(SharedMonsterAttributes.field_189429_h).func_111126_e())); //CM
+
             p_70665_2_ = this.func_70672_c(p_70665_1_, p_70665_2_);
             float f = p_70665_2_;
             p_70665_2_ = Math.max(p_70665_2_ - this.func_110139_bj(), 0.0F);
             this.func_110149_m(this.func_110139_bj() - (f - p_70665_2_));
+            DamageReporter.modify_damage(this, p_70665_1_, p_70665_2_, f, "Absorbtion"); //CM
 
             if (p_70665_2_ != 0.0F)
             {
                 float f1 = this.func_110143_aJ();
+                DamageReporter.register_final_damage(this, p_70665_1_, p_70665_2_); //CM
                 this.func_70606_j(f1 - p_70665_2_);
                 this.func_110142_aN().func_94547_a(p_70665_1_, f1, p_70665_2_);
                 this.func_110149_m(this.func_110139_bj() - p_70665_2_);
@@ -1828,7 +1875,10 @@
                             }
                         }
 
+                        if(carpet.carpetclient.CarpetClientChunkLogger.logger.enabled)
+                            carpet.carpetclient.CarpetClientChunkLogger.setReason("Entity walking around: " + func_70005_c_());
                         this.func_70091_d(MoverType.SELF, this.field_70159_w, this.field_70181_x, this.field_70179_y);
+                        carpet.carpetclient.CarpetClientChunkLogger.resetReason();
 
                         if (this.field_70123_F && this.func_70617_f_())
                         {
@@ -2306,6 +2356,9 @@
 
     protected void func_85033_bc()
     {
+        // Collision calculations are cancelled if canBePushed() is false. Return here to lag optimize the collision code. CARPET-XCOM
+        if(CarpetSettings.optimizedCollisionCancellations && !func_70104_M() && !(this instanceof net.minecraft.entity.boss.EntityDragon)) return;
+
         List<Entity> list = this.field_70170_p.func_175674_a(this, this.func_174813_aQ(), EntitySelectors.func_188442_a(this));
 
         if (!list.isEmpty())
@@ -2330,8 +2383,13 @@
                 }
             }
 
-            for (int l = 0; l < list.size(); ++l)
+            int limit = list.size();
+            if (CarpetSettings.maxEntityCollisions > 0)
             {
+                limit = Math.min(limit,CarpetSettings.maxEntityCollisions);
+            }
+            for (int l = 0; l < limit; ++l)
+            {
                 Entity entity = list.get(l);
                 this.func_82167_n(entity);
             }
@@ -2366,6 +2424,11 @@
     {
         this.field_70703_bu = p_70637_1_;
     }
+    //CM
+    public boolean getJumping()
+    {
+        return this.field_70703_bu;
+    }
 
     public void func_71001_a(Entity p_71001_1_, int p_71001_2_)
     {
