--- ../src-base/minecraft/net/minecraft/server/management/PlayerInteractionManager.java
+++ ../src-work/minecraft/net/minecraft/server/management/PlayerInteractionManager.java
@@ -8,6 +8,7 @@
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.entity.player.EntityPlayerMP;
+import net.minecraft.init.Blocks;
 import net.minecraft.inventory.IInventory;
 import net.minecraft.item.ItemBlock;
 import net.minecraft.item.ItemStack;
@@ -26,6 +27,10 @@
 import net.minecraft.world.World;
 import net.minecraft.world.WorldServer;
 
+import carpet.helpers.BlockRotator;
+import carpet.CarpetSettings;
+import carpet.worldedit.WorldEditBridge;
+
 public class PlayerInteractionManager
 {
     public World field_73092_a;
@@ -39,6 +44,7 @@
     private BlockPos field_180241_i = BlockPos.field_177992_a;
     private int field_73093_n;
     private int field_73094_o = -1;
+    public static EntityPlayerMP playerMinedBlock;
 
     public PlayerInteractionManager(World p_i1524_1_)
     {
@@ -137,6 +143,12 @@
 
     public void func_180784_a(BlockPos p_180784_1_, EnumFacing p_180784_2_)
     {
+        boolean worldEditResult = WorldEditBridge.onLeftClickBlock(field_73092_a, p_180784_1_, field_73090_b);
+        if (!worldEditResult)
+        {
+            field_73090_b.field_71135_a.func_147359_a(new SPacketBlockChange(field_73092_a, p_180784_1_));
+            return;
+        }
         if (this.func_73083_d())
         {
             if (!this.field_73092_a.func_175719_a((EntityPlayer)null, p_180784_1_, p_180784_2_))
@@ -191,6 +203,11 @@
                 this.field_73088_d = true;
                 this.field_180240_f = p_180784_1_;
                 int i = (int)(f * 10.0F);
+                //CM mining ghost blocks fix
+                if (CarpetSettings.miningGhostBlocksFix)
+                {
+                    field_73090_b.field_71135_a.func_147359_a(new SPacketBlockChange(field_73092_a, field_180240_f));
+                }
                 this.field_73092_a.func_175715_c(this.field_73090_b.func_145782_y(), p_180784_1_, i);
                 this.field_73094_o = i;
             }
@@ -235,8 +252,10 @@
     {
         IBlockState iblockstate = this.field_73092_a.func_180495_p(p_180235_1_);
         iblockstate.func_177230_c().func_176208_a(this.field_73092_a, p_180235_1_, iblockstate, this.field_73090_b);
+        if(carpet.carpetclient.CarpetClientChunkLogger.logger.enabled)
+            carpet.carpetclient.CarpetClientChunkLogger.setReason("Player removed block");
         boolean flag = this.field_73092_a.func_175698_g(p_180235_1_);
-
+        carpet.carpetclient.CarpetClientChunkLogger.resetReason();
         if (flag)
         {
             iblockstate.func_177230_c().func_176206_d(this.field_73092_a, p_180235_1_, iblockstate);
@@ -288,7 +307,15 @@
                 }
 
                 this.field_73092_a.func_180498_a(this.field_73090_b, 2001, p_180237_1_, Block.func_176210_f(iblockstate));
-                boolean flag1 = this.func_180235_c(p_180237_1_);
+                boolean flag1;
+                // Fix update suppression from not damaging tools or preventing dropping harvested blocks CARPET-XCOM
+                if(CarpetSettings.duplicationFixUpdateSuppression) {
+                    flag1 = field_73092_a.isRemovableBlock(p_180237_1_);
+                } else {
+                    playerMinedBlock = field_73090_b;
+                    flag1 = this.func_180235_c(p_180237_1_);
+                    playerMinedBlock = null;
+                }
 
                 if (this.func_73083_d())
                 {
@@ -307,10 +334,19 @@
 
                     if (flag1 && flag)
                     {
+                        playerMinedBlock = field_73090_b;
                         iblockstate.func_177230_c().func_180657_a(this.field_73092_a, this.field_73090_b, p_180237_1_, iblockstate, tileentity, itemstack2);
+                        playerMinedBlock = null;
                     }
                 }
 
+                // Update suppression duplication fix removing the block post inventory updates with exception to flowing water as ice can turn into flowing water CARPET-XCOM
+                if(CarpetSettings.duplicationFixUpdateSuppression && field_73092_a.func_180495_p(p_180237_1_) != Blocks.field_150358_i.func_176223_P()) {
+                    playerMinedBlock = field_73090_b;
+                    this.func_180235_c(p_180237_1_);
+                    playerMinedBlock = null;
+                }
+
                 return flag1;
             }
         }
@@ -402,10 +438,22 @@
         }
         else
         {
+            boolean worldEditResult = WorldEditBridge.onRightClickBlock(field_73092_a, p_187251_5_, this.field_73090_b);
+            if (!worldEditResult)
+            {
+                this.field_73090_b.field_71135_a.func_147359_a(new SPacketBlockChange(field_73092_a, p_187251_5_));
+                return EnumActionResult.FAIL;
+            }
             if (!p_187251_1_.func_70093_af() || p_187251_1_.func_184614_ca().func_190926_b() && p_187251_1_.func_184592_cb().func_190926_b())
             {
                 IBlockState iblockstate = p_187251_2_.func_180495_p(p_187251_5_);
 
+                //flip method will check for flippinCactus setting
+                Boolean flipped = BlockRotator.flipBlockWithCactus(p_187251_2_, p_187251_5_, iblockstate, p_187251_1_, p_187251_4_, p_187251_6_, p_187251_7_, p_187251_8_, p_187251_9_);
+                if (flipped)
+                {
+                    return EnumActionResult.PASS;
+                }
                 if (iblockstate.func_177230_c().func_180639_a(p_187251_2_, p_187251_5_, iblockstate, p_187251_1_, p_187251_4_, p_187251_6_, p_187251_7_, p_187251_8_, p_187251_9_))
                 {
                     return EnumActionResult.SUCCESS;
