--- ../src-base/minecraft/net/minecraft/server/management/PlayerChunkMap.java
+++ ../src-work/minecraft/net/minecraft/server/management/PlayerChunkMap.java
@@ -1,8 +1,10 @@
 package net.minecraft.server.management;
 
+import carpet.CarpetSettings;
 import com.google.common.base.Predicate;
 import com.google.common.collect.AbstractIterator;
 import com.google.common.collect.ComparisonChain;
+import com.google.common.collect.Iterators;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
@@ -215,9 +217,17 @@
 
             if (!worldprovider.func_76567_e())
             {
+                if(carpet.carpetclient.CarpetClientChunkLogger.logger.enabled)
+                    carpet.carpetclient.CarpetClientChunkLogger.setReason("Dimensional unloading due to no players");
                 this.field_72701_a.func_72863_F().func_73240_a();
+                carpet.carpetclient.CarpetClientChunkLogger.resetReason();
             }
         }
+
+        // Sends updates to all subscribed players that want to get indexing of chunks Carpet-XCOM
+        if(carpet.carpetclient.CarpetClientRandomtickingIndexing.sendUpdates(field_72701_a)) {
+            carpet.carpetclient.CarpetClientRandomtickingIndexing.sendRandomtickingChunkOrder(field_72701_a, this);
+        }
     }
 
     public boolean func_152621_a(int p_152621_1_, int p_152621_2_)
@@ -232,20 +242,21 @@
         return (PlayerChunkMapEntry)this.field_72700_c.get(func_187307_d(p_187301_1_, p_187301_2_));
     }
 
-    private PlayerChunkMapEntry func_187302_c(int p_187302_1_, int p_187302_2_)
+    private PlayerChunkMapEntry getOrCreateEntry(int chunkX, int chunkZ, EntityPlayerMP player)
     {
-        long i = func_187307_d(p_187302_1_, p_187302_2_);
+        long i = func_187307_d(chunkX, chunkZ);
         PlayerChunkMapEntry playerchunkmapentry = (PlayerChunkMapEntry)this.field_72700_c.get(i);
 
         if (playerchunkmapentry == null)
         {
-            playerchunkmapentry = new PlayerChunkMapEntry(this, p_187302_1_, p_187302_2_);
+            playerchunkmapentry = new PlayerChunkMapEntry(this, chunkX, chunkZ, player);
             this.field_72700_c.put(i, playerchunkmapentry);
             this.field_111193_e.add(playerchunkmapentry);
 
             if (playerchunkmapentry.func_187266_f() == null)
             {
-                this.field_187311_h.add(playerchunkmapentry);
+                // Added a way to remove spectators loading chunks. CARPET-XCOM
+                if(!CarpetSettings.spectatorsDontLoadChunks || !player.func_175149_v()) this.field_187311_h.add(playerchunkmapentry);
             }
 
             if (!playerchunkmapentry.func_187272_b())
@@ -271,8 +282,16 @@
 
     public void func_72683_a(EntityPlayerMP p_72683_1_)
     {
-        int i = (int)p_72683_1_.field_70165_t >> 4;
-        int j = (int)p_72683_1_.field_70161_v >> 4;
+        int i;
+        int j;
+        // Fix the player chunk map trunkation in negative coords causing offsets in chunk loading CARPET-XCOM
+        if(!CarpetSettings.playerChunkLoadingFix) {
+            i = (int)p_72683_1_.field_70165_t >> 4;
+            j = (int)p_72683_1_.field_70161_v >> 4;
+        } else {
+            i = MathHelper.func_76128_c(p_72683_1_.field_70165_t) >> 4;
+            j = MathHelper.func_76128_c(p_72683_1_.field_70161_v) >> 4;
+        }
         p_72683_1_.field_71131_d = p_72683_1_.field_70165_t;
         p_72683_1_.field_71132_e = p_72683_1_.field_70161_v;
 
@@ -280,7 +299,7 @@
         {
             for (int l = j - this.field_72698_e; l <= j + this.field_72698_e; ++l)
             {
-                this.func_187302_c(k, l).func_187276_a(p_72683_1_);
+                this.getOrCreateEntry(k, l, p_72683_1_).func_187276_a(p_72683_1_);
             }
         }
 
@@ -290,8 +309,16 @@
 
     public void func_72695_c(EntityPlayerMP p_72695_1_)
     {
-        int i = (int)p_72695_1_.field_71131_d >> 4;
-        int j = (int)p_72695_1_.field_71132_e >> 4;
+        int i;
+        int j;
+        // Fix the player chunk map trunkation in negative coords causing offsets in chunk loading CARPET-XCOM
+        if(!CarpetSettings.playerChunkLoadingFix) {
+            i = (int)p_72695_1_.field_71131_d >> 4;
+            j = (int)p_72695_1_.field_71132_e >> 4;
+        } else {
+            i = MathHelper.func_76128_c(p_72695_1_.field_71131_d) >> 4;
+            j = MathHelper.func_76128_c(p_72695_1_.field_71132_e) >> 4;
+        }
 
         for (int k = i - this.field_72698_e; k <= i + this.field_72698_e; ++k)
         {
@@ -327,16 +354,32 @@
 
     public void func_72685_d(EntityPlayerMP p_72685_1_)
     {
-        int i = (int)p_72685_1_.field_70165_t >> 4;
-        int j = (int)p_72685_1_.field_70161_v >> 4;
+        int i;
+        int j;
+        // Fix the player chunk map trunkation in negative coords causing offsets in chunk loading CARPET-XCOM
+        if(!CarpetSettings.playerChunkLoadingFix) {
+            i = (int)p_72685_1_.field_70165_t >> 4;
+            j = (int)p_72685_1_.field_70161_v >> 4;
+        } else {
+            i = MathHelper.func_76128_c(p_72685_1_.field_70165_t) >> 4;
+            j = MathHelper.func_76128_c(p_72685_1_.field_70161_v) >> 4;
+        }
         double d0 = p_72685_1_.field_71131_d - p_72685_1_.field_70165_t;
         double d1 = p_72685_1_.field_71132_e - p_72685_1_.field_70161_v;
         double d2 = d0 * d0 + d1 * d1;
 
         if (d2 >= 64.0D)
         {
-            int k = (int)p_72685_1_.field_71131_d >> 4;
-            int l = (int)p_72685_1_.field_71132_e >> 4;
+            int k;
+            int l;
+            // Fix the player chunk map trunkation in negative coords causing offsets in chunk loading CARPET-XCOM
+            if(!CarpetSettings.playerChunkLoadingFix) {
+                k = (int)p_72685_1_.field_71131_d >> 4;
+                l = (int)p_72685_1_.field_71132_e >> 4;
+            } else {
+                k = MathHelper.func_76128_c(p_72685_1_.field_71131_d) >> 4;
+                l = MathHelper.func_76128_c(p_72685_1_.field_71132_e) >> 4;
+            }
             int i1 = this.field_72698_e;
             int j1 = i - k;
             int k1 = j - l;
@@ -349,7 +392,7 @@
                     {
                         if (!this.func_72684_a(l1, i2, k, l, i1))
                         {
-                            this.func_187302_c(l1, i2).func_187276_a(p_72685_1_);
+                            this.getOrCreateEntry(l1, i2, p_72685_1_).func_187276_a(p_72685_1_);
                         }
 
                         if (!this.func_72684_a(l1 - j1, i2 - k1, i, j, i1))
@@ -396,7 +439,7 @@
                     {
                         for (int k1 = k - p_152622_1_; k1 <= k + p_152622_1_; ++k1)
                         {
-                            PlayerChunkMapEntry playerchunkmapentry = this.func_187302_c(j1, k1);
+                            PlayerChunkMapEntry playerchunkmapentry = this.getOrCreateEntry(j1, k1, entityplayermp);
 
                             if (!playerchunkmapentry.func_187275_d(entityplayermp))
                             {
@@ -413,7 +456,7 @@
                         {
                             if (!this.func_72684_a(l, i1, j, k, p_152622_1_))
                             {
-                                this.func_187302_c(l, i1).func_187277_b(entityplayermp);
+                                this.getOrCreateEntry(l, i1, entityplayermp).func_187277_b(entityplayermp);
                             }
                         }
                     }
@@ -460,7 +503,28 @@
 
         if (chunk != null)
         {
+            if(carpet.carpetclient.CarpetClientChunkLogger.logger.enabled)
+                carpet.carpetclient.CarpetClientChunkLogger.setReason("Player leaving chunk, queuing unload");
             this.func_72688_a().func_72863_F().func_189549_a(chunk);
+            carpet.carpetclient.CarpetClientChunkLogger.resetReason();
         }
     }
+    
+    /*
+     * 0x Chunk Logger - Gets the coordinates of all chunks
+     */
+    public Iterator<ChunkPos> carpetGetAllChunkCoordinates(){
+    	return new AbstractIterator<ChunkPos>() {
+    		Iterator<PlayerChunkMapEntry> allChunks = Iterators.concat(field_111193_e.iterator(),field_187311_h.iterator());
+			@Override
+			protected ChunkPos computeNext() {
+				if(allChunks.hasNext()) {
+					return allChunks.next().func_187264_a();
+				}
+				else {
+					return (ChunkPos) this.endOfData();
+				}
+			}
+    	};
+    }
 }
