--- ../src-base/minecraft/net/minecraft/server/management/PlayerChunkMapEntry.java
+++ ../src-work/minecraft/net/minecraft/server/management/PlayerChunkMapEntry.java
@@ -19,11 +19,14 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import carpet.carpetclient.CarpetClientChunkLogger;
+import carpet.helpers.IPlayerSensitiveTileEntity;
+
 public class PlayerChunkMapEntry
 {
     private static final Logger field_187281_a = LogManager.getLogger();
     private final PlayerChunkMap field_187282_b;
-    private final List<EntityPlayerMP> field_187283_c = Lists.<EntityPlayerMP>newArrayList();
+    public final List<EntityPlayerMP> field_187283_c = Lists.<EntityPlayerMP>newArrayList(); // CM: changed to public
     private final ChunkPos field_187284_d;
     private final short[] field_187285_e = new short[64];
     @Nullable
@@ -53,6 +56,11 @@
         }
         else
         {
+            // ChunkLogger - 0x-CARPET
+            if(CarpetClientChunkLogger.logger.enabled) {
+            	CarpetClientChunkLogger.logger.log(this.field_187282_b.func_72688_a(),field_187284_d.field_77276_a,field_187284_d.field_77275_b,CarpetClientChunkLogger.Event.PLAYER_ENTERS);
+            }
+
             if (this.field_187283_c.isEmpty())
             {
                 this.field_187289_i = this.field_187282_b.func_72688_a().func_82737_E();
@@ -80,6 +88,11 @@
 
             if (this.field_187283_c.isEmpty())
             {
+                // ChunkLogger - 0x-CARPET
+                if(CarpetClientChunkLogger.logger.enabled) {
+                	CarpetClientChunkLogger.logger.log(this.field_187282_b.func_72688_a(),field_187284_d.field_77276_a,field_187284_d.field_77275_b,CarpetClientChunkLogger.Event.PLAYER_LEAVES);
+                }
+            	
                 this.field_187282_b.func_187305_b(this);
             }
         }
@@ -248,12 +261,28 @@
     {
         if (p_187273_1_ != null)
         {
+            if (p_187273_1_ instanceof IPlayerSensitiveTileEntity)
+            {
+                if (field_187290_j)
+                {
+                    for (int i = 0; i < field_187283_c.size(); i++)
+                    {
+                        EntityPlayerMP player = field_187283_c.get(i);
+                        SPacketUpdateTileEntity packet = ((IPlayerSensitiveTileEntity) p_187273_1_).getUpdatePacketPlayerSensitive(player);
+                        if (packet != null)
+                            player.field_71135_a.func_147359_a(packet);
+                    }
+                }
+            }
+            else
+            {
             SPacketUpdateTileEntity spacketupdatetileentity = p_187273_1_.func_189518_D_();
 
             if (spacketupdatetileentity != null)
             {
                 this.func_187267_a(spacketupdatetileentity);
             }
+            }
         }
     }
 
