--- ../src-base/minecraft/net/minecraft/inventory/Container.java
+++ ../src-work/minecraft/net/minecraft/inventory/Container.java
@@ -17,6 +17,9 @@
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.World;
 
+import carpet.helpers.CtrlQcrafting;
+import carpet.CarpetSettings;
+
 public abstract class Container
 {
     public NonNullList<ItemStack> field_75153_a = NonNullList.<ItemStack>func_191196_a();
@@ -293,7 +296,7 @@
                         }
                         else if (slot6.func_75214_a(itemstack11))
                         {
-                            if (itemstack8.func_77973_b() == itemstack11.func_77973_b() && itemstack8.func_77960_j() == itemstack11.func_77960_j() && ItemStack.func_77970_a(itemstack8, itemstack11))
+                            if (itemstack8.func_77973_b() == itemstack11.func_77973_b() && itemstack8.func_77960_j() == itemstack11.func_77960_j() && ItemStack.func_77970_a(itemstack8, itemstack11) && itemstack11.func_77985_e()) // Check If item can stack, Always true in vanilla CARPET-XCOM
                             {
                                 int k2 = p_184996_2_ == 0 ? itemstack11.func_190916_E() : 1;
 
@@ -414,9 +417,22 @@
 
             if (slot2 != null && slot2.func_75216_d() && slot2.func_82869_a(p_184996_4_))
             {
-                ItemStack itemstack4 = slot2.func_75209_a(p_184996_2_ == 0 ? 1 : slot2.func_75211_c().func_190916_E());
-                slot2.func_190901_a(p_184996_4_, itemstack4);
-                p_184996_4_.func_71019_a(itemstack4, true);
+                /*
+                 * Adding Ctrl + Q to craft and throw. CARPET-XCOM
+                 */
+                if(CarpetSettings.ctrlQCraftingFix && p_184996_1_ == 0 && p_184996_2_ == 1)
+                {
+                    for (ItemStack itemstackDropAll = CtrlQcrafting.dropAllCrafting(p_184996_4_, p_184996_1_, field_75151_b); !itemstackDropAll.func_190926_b() && ItemStack.func_179545_c(slot2.func_75211_c(), itemstackDropAll); itemstackDropAll = CtrlQcrafting.dropAllCrafting(p_184996_4_, p_184996_1_, field_75151_b))
+                    {
+                        itemstack = itemstackDropAll.func_77946_l();
+                    }
+                }
+                else
+                {
+                    ItemStack itemstack4 = slot2.func_75209_a(p_184996_2_ == 0 ? 1 : slot2.func_75211_c().func_190916_E());
+                    slot2.func_190901_a(p_184996_4_, itemstack4);
+                    p_184996_4_.func_71019_a(itemstack4, true);
+                }
             }
         }
         else if (p_184996_3_ == ClickType.PICKUP_ALL && p_184996_1_ >= 0)
