--- ../src-base/minecraft/net/minecraft/inventory/ContainerWorkbench.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerWorkbench.java
@@ -140,4 +140,9 @@
     {
         return p_94530_2_.field_75224_c != this.field_75160_f && super.func_94530_a(p_94530_1_, p_94530_2_);
     }
+
+    // Method added for auto crafting CARPET-XCOM
+    public InventoryCrafting getInventoryCrafting() {
+        return field_75162_e;
+    }
 }
