--- ../src-base/minecraft/net/minecraft/village/VillageCollection.java
+++ ../src-work/minecraft/net/minecraft/village/VillageCollection.java
@@ -1,5 +1,6 @@
 package net.minecraft.village;
 
+import carpet.CarpetSettings;
 import com.google.common.collect.Lists;
 import java.util.Iterator;
 import java.util.List;
@@ -16,6 +17,8 @@
 import net.minecraft.world.WorldProvider;
 import net.minecraft.world.storage.WorldSavedData;
 
+import carpet.carpetclient.CarpetClientMarkers;
+
 public class VillageCollection extends WorldSavedData
 {
     private World field_75556_a;
@@ -69,7 +72,7 @@
         this.func_75549_c();
         this.func_75543_d();
         this.func_75545_e();
-
+        
         if (this.field_75553_e % 400 == 0)
         {
             this.func_76185_a();
