--- ../src-base/minecraft/net/minecraft/world/storage/SaveHandler.java
+++ ../src-work/minecraft/net/minecraft/world/storage/SaveHandler.java
@@ -7,6 +7,9 @@
 import java.io.FileOutputStream;
 import java.io.IOException;
 import javax.annotation.Nullable;
+
+import carpet.patches.EntityPlayerMPFake;
+import net.minecraft.entity.Entity;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.nbt.CompressedStreamTools;
 import net.minecraft.nbt.NBTTagCompound;
