--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIWatchClosest2.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIWatchClosest2.java
@@ -10,4 +10,9 @@
         super(p_i1629_1_, p_i1629_2_, p_i1629_3_, p_i1629_4_);
         this.func_75248_a(3);
     }
+
+    @Override
+    public String getTask() {
+        return ": " + field_75332_b.func_70005_c_() + " Task: Look at closest type 2";
+    }
 }
