--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIMoveTowardsTarget.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIMoveTowardsTarget.java
@@ -66,4 +66,9 @@
     {
         this.field_75431_a.func_70661_as().func_75492_a(this.field_75430_c, this.field_75427_d, this.field_75428_e, this.field_75425_f);
     }
+
+    @Override
+    public String getTask() {
+        return ": " + field_75431_a.func_70005_c_() + " Task: Moving towards target";
+    }
 }
