--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIRestrictSun.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIRestrictSun.java
@@ -27,4 +27,9 @@
     {
         ((PathNavigateGround)this.field_75273_a.func_70661_as()).func_179685_e(false);
     }
+
+    @Override
+    public String getTask() {
+        return ": " + field_75273_a.func_70005_c_() + " Task: Restrict sun";
+    }
 }
