--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIMoveIndoors.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIMoveIndoors.java
@@ -88,4 +88,9 @@
         this.field_75421_d = this.field_75422_b.func_179856_e().func_177952_p();
         this.field_75422_b = null;
     }
+
+    @Override
+    public String getTask() {
+        return ": " + field_75424_a.func_70005_c_() + " Task: Moving indoors";
+    }
 }
