--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIAttackRanged.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIAttackRanged.java
@@ -114,4 +114,9 @@
             this.field_75320_d = MathHelper.func_76141_d(f2 * (float)(this.field_75325_h - this.field_96561_g) + (float)this.field_96561_g);
         }
     }
+
+    @Override
+    public String getTask() {
+        return ": " + field_75322_b.func_70005_c_() + " Task: Ranged attack";
+    }
 }
