--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIWatchClosest.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIWatchClosest.java
@@ -90,4 +90,9 @@
         this.field_75332_b.func_70671_ap().func_75650_a(this.field_75334_a.field_70165_t, this.field_75334_a.field_70163_u + (double)this.field_75334_a.func_70047_e(), this.field_75334_a.field_70161_v, (float)this.field_75332_b.func_184649_cE(), (float)this.field_75332_b.func_70646_bf());
         --this.field_75330_d;
     }
+
+    @Override
+    public String getTask() {
+        return ": " + field_75332_b.func_70005_c_() + " Task: Look at closest";
+    }
 }
