--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIMoveToBlock.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIMoveToBlock.java
@@ -103,4 +103,9 @@
     }
 
     protected abstract boolean func_179488_a(World p_179488_1_, BlockPos p_179488_2_);
+
+    @Override
+    public String getTask() {
+        return ": " + field_179495_c.func_70005_c_() + " Task: Moving to block";
+    }
 }
