--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIFindEntityNearest.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIFindEntityNearest.java
@@ -117,4 +117,9 @@
         IAttributeInstance iattributeinstance = this.field_179442_b.func_110148_a(SharedMonsterAttributes.field_111265_b);
         return iattributeinstance == null ? 16.0D : iattributeinstance.func_111126_e();
     }
+
+    @Override
+    public String getTask() {
+        return ": " + field_179442_b.func_70005_c_() + " Task: Looking for nearest other entity";
+    }
 }
