--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIWanderAvoidWaterFlying.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIWanderAvoidWaterFlying.java
@@ -69,4 +69,9 @@
 
         return new Vec3d((double)blockpos1.func_177958_n(), (double)blockpos1.func_177956_o(), (double)blockpos1.func_177952_p());
     }
+
+    @Override
+    public String getTask() {
+        return ": " + field_75457_a.func_70005_c_() + " Task: Wander and avoid water while flying";
+    }
 }
