--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAILookIdle.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAILookIdle.java
@@ -38,4 +38,9 @@
         --this.field_75255_d;
         this.field_75258_a.func_70671_ap().func_75650_a(this.field_75258_a.field_70165_t + this.field_75256_b, this.field_75258_a.field_70163_u + (double)this.field_75258_a.func_70047_e(), this.field_75258_a.field_70161_v + this.field_75257_c, (float)this.field_75258_a.func_184649_cE(), (float)this.field_75258_a.func_70646_bf());
     }
+
+    @Override
+    public String getTask() {
+        return ": " + field_75258_a.func_70005_c_() + " Task: Looking idol";
+    }
 }
