--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIFollowParent.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIFollowParent.java
@@ -93,4 +93,9 @@
             this.field_75348_a.func_70661_as().func_75497_a(this.field_75346_b, this.field_75347_c);
         }
     }
+
+    @Override
+    public String getTask() {
+        return ": " + field_75348_a.func_70005_c_() + " Task: Following parrent";
+    }
 }
