--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIRunAroundLikeCrazy.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIRunAroundLikeCrazy.java
@@ -84,4 +84,9 @@
             this.field_111180_a.field_70170_p.func_72960_a(this.field_111180_a, (byte)6);
         }
     }
+
+    @Override
+    public String getTask() {
+        return ": " + field_111180_a.func_70005_c_() + " Task: Run around like crazy";
+    }
 }
