--- ../src-base/minecraft/net/minecraft/entity/item/EntityFallingBlock.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityFallingBlock.java
@@ -30,6 +30,10 @@
 import net.minecraft.util.math.Vec3d;
 import net.minecraft.world.World;
 
+import carpet.CarpetSettings;
+import carpet.logging.logHelpers.TrajectoryLogHelper;
+import carpet.logging.LoggerRegistry;
+
 public class EntityFallingBlock extends Entity
 {
     private IBlockState field_175132_d;
@@ -39,9 +43,24 @@
     private boolean field_145809_g;
     private int field_145815_h = 40;
     private float field_145816_i = 2.0F;
+
+    //CM
+    private int iceCount = 0;
+
     public NBTTagCompound field_145810_d;
     protected static final DataParameter<BlockPos> field_184532_d = EntityDataManager.<BlockPos>func_187226_a(EntityFallingBlock.class, DataSerializers.field_187200_j);
+    
+    // ----- Carpet Start ----- //
+    private TrajectoryLogHelper logHelper = null;
 
+    @Override
+    public void func_70106_y() {
+        if (LoggerRegistry.__fallingBlocks && logHelper != null)
+            logHelper.onFinish();
+        super.func_70106_y();
+    }
+    // ----- Carpet End ----- //
+
     public EntityFallingBlock(World p_i1706_1_)
     {
         super(p_i1706_1_);
@@ -61,12 +80,18 @@
         this.field_70167_r = p_i45848_4_;
         this.field_70166_s = p_i45848_6_;
         this.func_184530_a(new BlockPos(this));
+        // CM
+        this.iceCount = 0;
+        if (LoggerRegistry.__fallingBlocks)
+            logHelper = new TrajectoryLogHelper("fallingBlocks");
     }
 
     public boolean func_70075_an()
     {
         return false;
     }
+    //CM
+    public String cm_name() { return "Falling Block"; }
 
     public void func_184530_a(BlockPos p_184530_1_)
     {
@@ -90,6 +115,11 @@
 
     public void func_70071_h_()
     {
+        // ----- Carpet Start ----- //
+        if (LoggerRegistry.__fallingBlocks && logHelper != null)
+            logHelper.onTick(field_70165_t, field_70163_u, field_70161_v, field_70159_w, field_70181_x, field_70179_y);
+        // ----- Carpet End ----- //
+
         Block block = this.field_175132_d.func_177230_c();
 
         if (this.field_175132_d.func_185904_a() == Material.field_151579_a)
@@ -122,8 +152,44 @@
                 this.field_70181_x -= 0.03999999910593033D;
             }
 
-            this.func_70091_d(MoverType.SELF, this.field_70159_w, this.field_70181_x, this.field_70179_y);
+            // Optimized TNT movement skipping the move code given its expensive if identical tnt movement is done. CARPET-XCOM
+            if(!CarpetSettings.fallingMovementOptimization) {
+                this.func_70091_d(MoverType.SELF, this.field_70159_w, this.field_70181_x, this.field_70179_y);
+            } else {
+                if(!cacheMatching()) {
+                    cache[0] = field_70165_t;
+                    cache[1] = field_70163_u;
+                    cache[2] = field_70161_v;
+                    cache[3] = field_70159_w;
+                    cache[4] = field_70181_x;
+                    cache[5] = field_70179_y;
+                    cacheTime = func_184102_h().func_71259_af();
+                    this.func_70091_d(MoverType.SELF, this.field_70159_w, this.field_70181_x, this.field_70179_y);
+                    if (!field_70128_L) {
+                        cache[6] = field_70165_t;
+                        cache[7] = field_70163_u;
+                        cache[8] = field_70161_v;
+                        cache[9] = field_70159_w;
+                        cache[10] = field_70181_x;
+                        cache[11] = field_70179_y;
+                        cacheBool[0] = field_70134_J;
+                        cacheBool[1] = field_70122_E;
+                    } else {
+                        cache[0] = Integer.MAX_VALUE;
+                    }
+                } else {
+                    this.func_70107_b(cache[6], cache[7], cache[8]);
+                    field_70159_w = cache[9];
+                    field_70181_x = cache[10];
+                    field_70179_y = cache[11];
+                    field_70134_J = cacheBool[0];
+                    field_70122_E = cacheBool[1];
+                }
+            }
 
+            //Fix falling blocks duplication when going through portals. CARPET-XCOM
+            if(CarpetSettings.duplicationFixGravityBlocks && field_70128_L) return;
+            
             if (!this.field_70170_p.field_72995_K)
             {
                 BlockPos blockpos1 = new BlockPos(this);
@@ -168,6 +234,24 @@
                     this.field_70179_y *= 0.699999988079071D;
                     this.field_70181_x *= -0.5D;
 
+                    if (block == Blocks.field_150467_bQ &&
+                            CarpetSettings.renewablePackedIce &&
+                            this.field_70170_p.func_180495_p(new BlockPos(this.field_70165_t, this.field_70163_u - 0.059999999776482582D, this.field_70161_v)).func_177230_c()==Blocks.field_150432_aD  )
+                    {
+                        if (iceCount < 2)
+                        {
+                            field_70170_p.func_175655_b(blockpos1.func_177977_b(), false);
+                            this.field_70122_E = false;
+                            iceCount++;
+                            return;
+                        }
+                        else
+                        {
+                            field_70170_p.func_180501_a(blockpos1.func_177977_b(),Blocks.field_150403_cj.func_176223_P(),3);
+                            field_70170_p.func_175718_b(2001,blockpos1.func_177977_b(),Block.func_149682_b(Blocks.field_150403_cj));
+                        }
+                    }
+
                     if (iblockstate.func_177230_c() != Blocks.field_180384_M)
                     {
                         this.func_70106_y();
@@ -356,4 +440,12 @@
     {
         return true;
     }
+
+    // Optimization methods CARPET-XCOM
+    private static double[] cache = new double[12];
+    private static boolean[] cacheBool = new boolean[2];
+    private static long cacheTime = 0;
+    private boolean cacheMatching() {
+        return cache[0] == field_70165_t && cache[1] == field_70163_u && cache[2] == field_70161_v && cache[3] == field_70159_w && cache[4] == field_70181_x && cache[5] == field_70179_y && cacheTime == func_184102_h().func_71259_af();
+    }
 }
