--- ../src-base/minecraft/net/minecraft/entity/item/EntityXPOrb.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityXPOrb.java
@@ -14,16 +14,24 @@
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.World;
 
+import carpet.CarpetSettings;
+import carpet.helpers.XPcombine;
+
 public class EntityXPOrb extends Entity
 {
     public int field_70533_a;
     public int field_70531_b;
     public int field_70532_c;
+    public int delayBeforeCombine;
+
     private int field_70529_d = 5;
-    private int field_70530_e;
+    public int field_70530_e;
     private EntityPlayer field_80001_f;
     private int field_80002_g;
 
+    //CM
+    public String cm_name() { return "XP Orb"; }
+
     public EntityXPOrb(World p_i1585_1_, double p_i1585_2_, double p_i1585_4_, double p_i1585_6_, int p_i1585_8_)
     {
         super(p_i1585_1_);
@@ -34,8 +42,26 @@
         this.field_70181_x = (double)((float)(Math.random() * 0.2D) * 2.0F);
         this.field_70179_y = (double)((float)(Math.random() * 0.20000000298023224D - 0.10000000149011612D) * 2.0F);
         this.field_70530_e = p_i1585_8_;
+        // add combine delay CARPET-XCOM
+        this.delayBeforeCombine = 50;
     }
 
+    /*
+     * New constructor for updating XP orb size change CARPET-XCOM
+     */
+    public EntityXPOrb(World worldIn, int expValue, EntityXPOrb old)
+    {
+        super(worldIn);
+        this.func_70105_a(0.5F, 0.5F);
+        this.func_70107_b(old.field_70165_t, old.field_70163_u, old.field_70161_v);
+        this.field_70177_z = old.field_70177_z;
+        this.field_70159_w = old.field_70159_w;
+        this.field_70181_x = old.field_70181_x;
+        this.field_70179_y = old.field_70179_y;
+        this.field_70530_e = expValue;
+        this.delayBeforeCombine = 50;
+    }
+
     protected boolean func_70041_e_()
     {
         return false;
@@ -113,6 +139,20 @@
         }
 
         this.func_70091_d(MoverType.SELF, this.field_70159_w, this.field_70181_x, this.field_70179_y);
+
+        /*
+         * Adding code here for combining XP orbs. CARPET-XCOM
+         */
+        if (CarpetSettings.combineXPOrbs )
+        {
+            if (this.delayBeforeCombine > 0)
+            {
+                --this.delayBeforeCombine;
+            }
+
+            XPcombine.searchForOtherXPNearbyCarpet(this);
+        }
+
         float f = 0.98F;
 
         if (this.field_70122_E)
@@ -186,7 +226,8 @@
     {
         if (!this.field_70170_p.field_72995_K)
         {
-            if (this.field_70532_c == 0 && p_70100_1_.field_71090_bL == 0)
+            //CM added ORed condition
+            if (this.field_70532_c == 0 && (p_70100_1_.field_71090_bL == 0 || CarpetSettings.xpNoCooldown) )
             {
                 p_70100_1_.field_71090_bL = 2;
                 p_70100_1_.func_71001_a(this, 1);
