--- ../src-base/minecraft/net/minecraft/entity/item/EntityItemFrame.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityItemFrame.java
@@ -23,6 +23,8 @@
 import net.minecraft.world.World;
 import net.minecraft.world.storage.MapData;
 
+import carpet.CarpetSettings;
+
 public class EntityItemFrame extends EntityHanging
 {
     private static final DataParameter<ItemStack> field_184525_c = EntityDataManager.<ItemStack>func_187226_a(EntityItemFrame.class, DataSerializers.field_187196_f);
@@ -97,7 +99,8 @@
 
     public void func_146065_b(@Nullable Entity p_146065_1_, boolean p_146065_2_)
     {
-        if (this.field_70170_p.func_82736_K().func_82766_b("doEntityDrops"))
+        // Added a fix for duplication of items when breaking item frames CARPET-XCOM
+        if (this.field_70170_p.func_82736_K().func_82766_b("doEntityDrops") && (!CarpetSettings.duplicationFixItemFrame || !field_70128_L))
         {
             ItemStack itemstack = this.func_82335_i();
 
