--- ../src-base/minecraft/net/minecraft/entity/item/EntityItem.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityItem.java
@@ -1,6 +1,12 @@
 package net.minecraft.entity.item;
 
 import javax.annotation.Nullable;
+
+import carpet.CarpetSettings;
+import carpet.helpers.HopperCounter;
+import carpet.logging.LoggerRegistry;
+import carpet.logging.logHelpers.ItemLogHelper;
+import carpet.utils.extensions.ExtendedItemStack;
 import net.minecraft.block.material.Material;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.MoverType;
@@ -35,6 +41,7 @@
     private String field_145801_f;
     private String field_145802_g;
     public float field_70290_d;
+    private ItemLogHelper logHelper = null;
 
     public EntityItem(World p_i1709_1_, double p_i1709_2_, double p_i1709_4_, double p_i1709_6_)
     {
@@ -47,6 +54,12 @@
         this.field_70159_w = (double)((float)(Math.random() * 0.20000000298023224D - 0.10000000149011612D));
         this.field_70181_x = 0.20000000298023224D;
         this.field_70179_y = (double)((float)(Math.random() * 0.20000000298023224D - 0.10000000149011612D));
+
+        //CM
+        if (LoggerRegistry.__items)
+        {
+            logHelper = new ItemLogHelper("items");
+        }
     }
 
     public EntityItem(World p_i1710_1_, double p_i1710_2_, double p_i1710_4_, double p_i1710_6_, ItemStack p_i1710_8_)
@@ -165,8 +178,20 @@
                 }
             }
 
+            // ----- Carpet Start ----- //
+            if (LoggerRegistry.__items && logHelper != null)
+            {
+                logHelper.onTick(field_70165_t, field_70163_u, field_70161_v, field_70159_w, field_70181_x, field_70179_y);
+            }
+            // ----- Carpet End ----- //
+
             if (!this.field_70170_p.field_72995_K && this.field_70292_b >= 6000)
             {
+                // ----- Carpet Start ----- //
+                if (LoggerRegistry.__items && logHelper != null) {
+                    logHelper.onFinish("Despawn Timer");
+                }
+                // ----- Carpet End ----- //
                 this.func_70106_y();
             }
         }
@@ -221,8 +246,23 @@
                     }
                     else if (itemstack1.func_190916_E() + itemstack.func_190916_E() > itemstack1.func_77976_d())
                     {
+                        // Add check for stacking shoulkers without NBT on the ground CARPET-XCOM
+                        if (((ExtendedItemStack) (Object) itemstack1).isGroundStackable() && ((ExtendedItemStack) (Object) itemstack).isGroundStackable())
+                        {
+                            itemstack1.func_190917_f(itemstack.func_190916_E());
+                            p_70289_1_.field_145804_b = Math.max(p_70289_1_.field_145804_b, this.field_145804_b);
+                            p_70289_1_.field_70292_b = Math.min(p_70289_1_.field_70292_b, this.field_70292_b);
+                            p_70289_1_.func_92058_a(itemstack1);
+                            this.func_70106_y();
+                            return true;
+                        }
                         return false;
                     }
+                    // make sure stackable items are checked before combining them, always true in vanilla CARPET-XCOM
+                    else if (!itemstack1.func_77985_e() && !itemstack.func_77985_e())
+                    {
+                        return false;
+                    }
                     else
                     {
                         itemstack1.func_190917_f(itemstack.func_190916_E());
@@ -295,6 +335,15 @@
 
             if (this.field_70291_e <= 0)
             {
+                // Add cactus counter CARPET-XCOM
+                if(CarpetSettings.cactusCounter && p_70097_1_ == DamageSource.field_76367_g){
+                    HopperCounter.cactus.add(field_70170_p.func_73046_m(), func_92059_d());
+                }
+                // ----- Carpet Start ----- //
+                if (LoggerRegistry.__items && logHelper != null) {
+                    logHelper.onFinish(p_70097_1_.func_76355_l());
+                }
+                // ----- Carpet End ----- //
                 this.func_70106_y();
             }
 
