--- ../src-base/minecraft/net/minecraft/world/gen/ChunkProviderServer.java
+++ ../src-work/minecraft/net/minecraft/world/gen/ChunkProviderServer.java
@@ -27,13 +27,18 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import net.minecraft.nbt.NBTTagList;
+import carpet.CarpetSettings;
+import carpet.utils.TickingArea;
+import net.minecraft.entity.Entity;
+
 public class ChunkProviderServer implements IChunkProvider
 {
     private static final Logger field_147417_b = LogManager.getLogger();
-    private final Set<Long> field_73248_b = Sets.<Long>newHashSet();
-    private final IChunkGenerator field_186029_c;
-    private final IChunkLoader field_73247_e;
-    private final Long2ObjectMap<Chunk> field_73244_f = new Long2ObjectOpenHashMap<Chunk>(8192);
+    public final Set<Long> field_73248_b = Sets.<Long>newHashSet();
+    public final IChunkGenerator field_186029_c;
+    public final IChunkLoader field_73247_e; // CM changed to public for ticking areas
+    public final Long2ObjectMap<Chunk> field_73244_f = new Long2ObjectOpenHashMap<Chunk>(8192); // CM changed to public for commandUnload
     private final WorldServer field_73251_h;
 
     public ChunkProviderServer(WorldServer p_i46838_1_, IChunkLoader p_i46838_2_, IChunkGenerator p_i46838_3_)
@@ -50,7 +55,12 @@
 
     public void func_189549_a(Chunk p_189549_1_)
     {
-        if (this.field_73251_h.field_73011_w.func_186056_c(p_189549_1_.field_76635_g, p_189549_1_.field_76647_h))
+        boolean canDrop = field_73251_h.field_73011_w.func_186056_c(p_189549_1_.field_76635_g, p_189549_1_.field_76647_h);
+        if (CarpetSettings.getBool("disableSpawnChunks"))
+            canDrop = true;
+        if (CarpetSettings.getBool("tickingAreas"))
+            canDrop &= !TickingArea.isTickingChunk(field_73251_h, p_189549_1_.field_76635_g, p_189549_1_.field_76647_h);
+        if (canDrop)
         {
             this.field_73248_b.add(Long.valueOf(ChunkPos.func_77272_a(p_189549_1_.field_76635_g, p_189549_1_.field_76647_h)));
             p_189549_1_.field_189550_d = true;
@@ -185,6 +195,8 @@
 
     public boolean func_186027_a(boolean p_186027_1_)
     {
+        // NewLight PHIPRO-CARPET
+        if (CarpetSettings.newLight) this.field_73251_h.lightingEngine.procLightUpdates();
         int i = 0;
         List<Chunk> list = Lists.newArrayList(this.field_73244_f.values());
 
@@ -224,6 +236,8 @@
         {
             if (!this.field_73248_b.isEmpty())
             {
+                // NewLight PHIPRO-CARPET
+                if (CarpetSettings.newLight) this.field_73251_h.lightingEngine.procLightUpdates();
                 Iterator<Long> iterator = this.field_73248_b.iterator();
 
                 for (int i = 0; i < 100 && iterator.hasNext(); iterator.remove())
@@ -288,4 +302,7 @@
     {
         return this.field_73244_f.containsKey(ChunkPos.func_77272_a(p_191062_1_, p_191062_2_)) || this.field_73247_e.func_191063_a(p_191062_1_, p_191062_2_);
     }
+
+    // Retrieval method to get the bounding boxes CARPET-XCOM
+    public NBTTagList getBoundingBoxes(Entity entity) { return this.field_186029_c.getBoundingBoxes(entity); }
 }
