--- ../src-base/minecraft/net/minecraft/world/gen/ChunkGeneratorEnd.java
+++ ../src-work/minecraft/net/minecraft/world/gen/ChunkGeneratorEnd.java
@@ -22,6 +22,8 @@
 import net.minecraft.world.gen.feature.WorldGenEndIsland;
 import net.minecraft.world.gen.structure.MapGenEndCity;
 
+import carpet.CarpetSettings;
+
 public class ChunkGeneratorEnd implements IChunkGenerator
 {
     private final Random field_73220_k;
@@ -420,5 +422,7 @@
 
     public void func_180514_a(Chunk p_180514_1_, int p_180514_2_, int p_180514_3_)
     {
+        //I assume its for shulker spaning in end cities
+        if (CarpetSettings.shulkerSpawningInEndCities) this.field_185972_n.func_186125_a(this.field_73230_p, p_180514_2_, p_180514_3_, (ChunkPrimer)null);
     }
 }
