--- ../src-base/minecraft/net/minecraft/world/Explosion.java
+++ ../src-work/minecraft/net/minecraft/world/Explosion.java
@@ -26,20 +26,26 @@
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.math.Vec3d;
 
+import carpet.CarpetSettings;
+import carpet.helpers.OptimizedTNT;
+
 public class Explosion
 {
-    private final boolean field_77286_a;
-    private final boolean field_82755_b;
-    private final Random field_77290_i = new Random();
-    private final World field_77287_j;
-    private final double field_77284_b;
-    private final double field_77285_c;
-    private final double field_77282_d;
-    private final Entity field_77283_e;
-    private final float field_77280_f;
-    private final List<BlockPos> field_77281_g = Lists.<BlockPos>newArrayList();
-    private final Map<EntityPlayer, Vec3d> field_77288_k = Maps.<EntityPlayer, Vec3d>newHashMap();
+    public final boolean field_77286_a;
+    public final boolean field_82755_b;
+    public final Random field_77290_i = new Random();
+    public final World field_77287_j;
+    public final double field_77284_b;
+    public final double field_77285_c;
+    public final double field_77282_d;
+    public final Entity field_77283_e;
+    public final float field_77280_f;
+    public final List<BlockPos> field_77281_g = Lists.<BlockPos>newArrayList();
+    public final Map<EntityPlayer, Vec3d> field_77288_k = Maps.<EntityPlayer, Vec3d>newHashMap();
 
+    // For disabling the explosion particles and sound CARPET-XCOM
+    public static int explosionSound = 0;
+
     public Explosion(World p_i45754_1_, Entity p_i45754_2_, double p_i45754_3_, double p_i45754_5_, double p_i45754_7_, float p_i45754_9_, boolean p_i45754_10_, boolean p_i45754_11_)
     {
         this.field_77287_j = p_i45754_1_;
@@ -54,6 +60,16 @@
 
     public void func_77278_a()
     {
+        /*
+         * Divert into a more optimized explosion code.
+         * CARPET-XCOM
+         */
+        if(CarpetSettings.optimizedTNT){
+            OptimizedTNT.doExplosionA(this);
+            return;
+        }
+
+        explosionSound = 0;
         Set<BlockPos> set = Sets.<BlockPos>newHashSet();
         int i = 16;
 
@@ -102,7 +118,7 @@
             }
         }
 
-        this.field_77281_g.addAll(set);
+        if(!CarpetSettings.explosionNoBlockDamage) this.field_77281_g.addAll(set);
         float f3 = this.field_77280_f * 2.0F;
         int k1 = MathHelper.func_76128_c(this.field_77284_b - (double)f3 - 1.0D);
         int l1 = MathHelper.func_76128_c(this.field_77284_b + (double)f3 + 1.0D);
@@ -164,6 +180,14 @@
 
     public void func_77279_a(boolean p_77279_1_)
     {
+        /*
+         * Divert into a more optimized explosion code.
+         * CARPET-MASA
+         */
+        if (CarpetSettings.optimizedTNT) {
+            OptimizedTNT.doExplosionB(this, p_77279_1_);
+            return;
+        }
         this.field_77287_j.func_184148_a((EntityPlayer)null, this.field_77284_b, this.field_77285_c, this.field_77282_d, SoundEvents.field_187539_bB, SoundCategory.BLOCKS, 4.0F, (1.0F + (this.field_77287_j.field_73012_v.nextFloat() - this.field_77287_j.field_73012_v.nextFloat()) * 0.2F) * 0.7F);
 
         if (this.field_77280_f >= 2.0F && this.field_82755_b)
