--- ../src-base/minecraft/net/minecraft/world/WorldEntitySpawner.java
+++ ../src-work/minecraft/net/minecraft/world/WorldEntitySpawner.java
@@ -22,6 +22,12 @@
 import net.minecraft.world.biome.Biome;
 import net.minecraft.world.chunk.Chunk;
 
+import net.minecraft.entity.EntityList;
+import carpet.utils.SpawnReporter;
+import carpet.CarpetSettings;
+import java.lang.Math;
+import net.minecraft.util.Tuple;
+
 public final class WorldEntitySpawner
 {
     private static final int field_180268_a = (int)Math.pow(17.0D, 2.0D);
@@ -72,19 +78,54 @@
                 }
             }
 
+            //CM start
+            boolean optimizedDespawnRange = CarpetSettings.optimizedDespawnRange;
+            if (i==0 && optimizedDespawnRange) // worlds without valid chunks are skipped.
+            {
+                return 0;
+            }
+            //CM end
             int j4 = 0;
             BlockPos blockpos1 = p_77192_1_.func_175694_M();
+            //CM start
+            int did = p_77192_1_.field_73011_w.func_186058_p().func_186068_a();
+            String level_suffix = (did==0)?"":((did<0?" (N)":" (E)"));
+            //CM end
 
             for (EnumCreatureType enumcreaturetype : EnumCreatureType.values())
             {
+                //CM start
+                String type_code = String.format("%s", enumcreaturetype);
+                String group_code = type_code+level_suffix;
+                if (SpawnReporter.track_spawns > 0L)
+                {
+                    SpawnReporter.overall_spawn_ticks.put(group_code, SpawnReporter.overall_spawn_ticks.get(group_code) + SpawnReporter.spawn_tries.get(type_code));
+                }
+                //CM end
                 if ((!enumcreaturetype.func_75599_d() || p_77192_3_) && (enumcreaturetype.func_75599_d() || p_77192_2_) && (!enumcreaturetype.func_82705_e() || p_77192_4_))
                 {
                     int k4 = p_77192_1_.func_72907_a(enumcreaturetype.func_75598_a());
-                    int l4 = enumcreaturetype.func_75601_b() * i / field_180268_a;
-
+                    //CM replaced: //int l4 = enumcreaturetype.getMaxNumberOfCreature() * i / MOB_COUNT_DIV;
+                    int l4 = (int)(Math.pow(2.0,(SpawnReporter.mobcap_exponent/4)) * enumcreaturetype.func_75601_b() * i / field_180268_a);
+                    SpawnReporter.mobcaps.get(did).put(enumcreaturetype, new Tuple<>(k4, l4));
+                    int tries = SpawnReporter.spawn_tries.get(type_code);
+                    if (SpawnReporter.track_spawns > 0L)
+                    {
+                        SpawnReporter.spawn_attempts.put(group_code, SpawnReporter.spawn_attempts.get(group_code) + tries);
+                        SpawnReporter.spawn_cap_count.put(group_code, SpawnReporter.spawn_cap_count.get(group_code) + k4);
+                    }
+                    if (SpawnReporter.mock_spawns) { k4 = 0; } // no mobcaps
+                    //CM end
+                    
                     if (k4 <= l4)
                     {
                         BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
+                      /* carpet mod -> extra indentation */
+                      for (int trie = 0; trie < tries; trie++)
+                      {
+                        long local_spawns = 0;
+                        /* end */
+                        
                         label134:
 
                         for (ChunkPos chunkpos1 : this.field_77193_b)
@@ -108,6 +149,8 @@
                                     Biome.SpawnListEntry biome$spawnlistentry = null;
                                     IEntityLivingData ientitylivingdata = null;
                                     int l3 = MathHelper.func_76143_f(Math.random() * 4.0D);
+                                    //CM fixed 4 mobs per pack
+                                    if (CarpetSettings._1_8Spawning) { l3 = 4; }
 
                                     for (int i4 = 0; i4 < l3; ++i4)
                                     {
@@ -153,7 +196,29 @@
                                                     if (entityliving.func_70058_J())
                                                     {
                                                         ++j2;
-                                                        p_77192_1_.func_72838_d(entityliving);
+                                                        //CM replacing //worldServerIn.spawnEntityInWorld(entityliving);
+                                                        if (optimizedDespawnRange && entityliving.willImmediatelyDespawn()) // Added optimized despawn mobs causing netlag by Luflosi CARPET-XCOM
+                                                        {
+                                                            entityliving.func_70106_y();
+                                                        }
+                                                        else
+                                                        {
+                                                            ++local_spawns;
+                                                            if (SpawnReporter.track_spawns > 0L)
+                                                            {
+                                                                String species = EntityList.func_75621_b(entityliving);
+                                                                SpawnReporter.registerSpawn(entityliving, type_code, species, blockpos$mutableblockpos);
+                                                            }
+                                                            if (SpawnReporter.mock_spawns)
+                                                            {
+                                                                entityliving.func_70106_y();
+                                                            }
+                                                            else
+                                                            {
+                                                                p_77192_1_.func_72838_d(entityliving);
+                                                            }
+                                                            //CM end
+                                                        }
                                                     }
                                                     else
                                                     {
@@ -173,7 +238,30 @@
                                 }
                             }
                         }
+                        /* carpet mod */
+                        
+                        if (SpawnReporter.track_spawns > 0L)
+                        {
+                            if (local_spawns > 0)
+                            {
+                                SpawnReporter.spawn_ticks_succ.put(group_code, SpawnReporter.spawn_ticks_succ.get(group_code) + 1L);
+                                SpawnReporter.spawn_ticks_spawns.put(group_code, SpawnReporter.spawn_ticks_spawns.get(group_code) + local_spawns);
+                            }
+                            else
+                            {
+                                SpawnReporter.spawn_ticks_fail.put(group_code, SpawnReporter.spawn_ticks_fail.get(group_code) + 1L);
+                            }
+                        }
+                      } //carpet mod <- extra indentation
                     }
+                    else //carpet mod full mobcap
+                    {
+                        if (SpawnReporter.track_spawns > 0L)
+                        {
+                            SpawnReporter.spawn_ticks_full.put(group_code, SpawnReporter.spawn_ticks_full.get(group_code) + SpawnReporter.spawn_tries.get(type_code));
+                        }
+                    }
+                    /* end */
                 }
             }
 
