--- ../src-base/minecraft/net/minecraft/world/World.java
+++ ../src-work/minecraft/net/minecraft/world/World.java
@@ -60,6 +60,9 @@
 import net.minecraft.world.storage.WorldSavedData;
 import net.minecraft.world.storage.loot.LootTableManager;
 
+import java.lang.reflect.Field;
+import java.util.concurrent.atomic.AtomicLong;
+
 public abstract class World implements IBlockAccess
 {
     private int field_181546_a = 63;
@@ -3361,4 +3364,33 @@
     {
         return null;
     }
+
+    AtomicLong scrambledSeed;
+    public long getRandSeed(){
+        try
+        {
+            if(scrambledSeed == null) {
+                Field field = Random.class.getDeclaredField("seed");
+                field.setAccessible(true);
+                scrambledSeed = (AtomicLong) field.get(field_73012_v);   //this needs to be XOR'd with 0x5DEECE66DL
+            }
+            return scrambledSeed.get();
+            // Minecraft.getMinecraft().player.sendChatMessage(chunk.x + ", " + chunk.z + ", seed " + theSeed);
+        } catch (Exception e) {}
+
+        return 0;
+    }
+
+    // Method to check if a block can be removed from the world CARPET-XCOM
+    public boolean isRemovableBlock(BlockPos pos) {
+        if (this.func_189509_E(pos)) {
+            return false;
+        } else if (!this.field_72995_K && this.field_72986_A.func_76067_t() == WorldType.field_180272_g) {
+            return false;
+        } else if (func_180495_p(pos) == Blocks.field_150350_a.func_176223_P()) {
+            return false;
+        }
+
+        return true;
+    }
 }
