--- ../src-base/minecraft/net/minecraft/world/WorldServer.java
+++ ../src-work/minecraft/net/minecraft/world/WorldServer.java
@@ -80,6 +80,8 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import carpet.helpers.RandomTickOptimization;
+
 public class WorldServer extends World implements IThreadListener
 {
     private static final Logger field_147491_a = LogManager.getLogger();
@@ -503,9 +505,10 @@
                     {
                         iblockstate.func_177230_c().func_180650_b(this, p_175654_1_, iblockstate, this.field_73012_v);
                     }
+                    if(RandomTickOptimization.needsWorldGenFix) return;
                 }
 
-                return;
+                if(!RandomTickOptimization.needsWorldGenFix) return;
             }
 
             p_175654_3_ = 1;
