--- ../src-base/minecraft/net/minecraft/entity/monster/EntityGhast.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityGhast.java
@@ -29,6 +29,13 @@
 import net.minecraft.world.World;
 import net.minecraft.world.storage.loot.LootTableList;
 
+import carpet.CarpetSettings;
+import net.minecraft.util.EnumHand;
+import net.minecraft.entity.Entity;
+import net.minecraft.init.Items;
+import com.google.common.base.Predicate;
+import carpet.helpers.GhastHelper;
+
 public class EntityGhast extends EntityFlying implements IMob
 {
     private static final DataParameter<Boolean> field_184683_a = EntityDataManager.<Boolean>func_187226_a(EntityGhast.class, DataSerializers.field_187198_h);
@@ -48,7 +55,9 @@
         this.field_70714_bg.func_75776_a(5, new EntityGhast.AIRandomFly(this));
         this.field_70714_bg.func_75776_a(7, new EntityGhast.AILookAround(this));
         this.field_70714_bg.func_75776_a(7, new EntityGhast.AIFireballAttack(this));
-        this.field_70715_bh.func_75776_a(1, new EntityAIFindEntityNearestPlayer(this));
+        this.field_70714_bg.func_75776_a(3, new GhastHelper.AIFollowClues(this)); //[CM rideableGhasts]
+        this.field_70714_bg.func_75776_a(4, new GhastHelper.AIFindOwner(this)); //[CM rideableGhasts]
+        this.field_70715_bh.func_75776_a(1, new GhastHelper.GhastEntityAIFindEntityNearestPlayer(this)); //CM modified base AI task
     }
 
     public void func_175454_a(boolean p_175454_1_)
@@ -168,6 +177,47 @@
         return 2.6F;
     }
 
+    //[CM]
+    public void func_70636_d()
+    {
+        super.func_70636_d();
+        if (this.field_70146_Z.nextInt(400) == 0 && CarpetSettings.rideableGhasts && this.field_70725_aQ == 0 && this.func_184207_aI())
+        {
+            this.func_70691_i(1.0F);
+        }
+    }
+
+    /* Carpet mod rideable Ghasts
+    * rightclick action
+    * returning false propagates other checks like nametagging
+    */
+    public boolean func_184645_a(EntityPlayer player, EnumHand hand)
+    {
+        if (!(CarpetSettings.rideableGhasts) || this.func_184207_aI())
+        {
+            return super.func_184645_a(player, hand);
+        }
+        if (!GhastHelper.is_yo_bro(this, player))
+        {
+            return super.func_184645_a(player, hand);
+        }
+        boolean worked = super.func_184645_a(player, hand);
+        if (!worked)
+        {
+            player.func_184205_a(this, true);
+            player.func_184811_cZ().func_185145_a(Items.field_151059_bz, 1);
+        }
+        return false;
+
+    }
+    /*signify that its a rideable thingy - not sure if it is all needed
+    */
+    public Entity func_184179_bs()
+    {
+        return this.func_184188_bt().isEmpty() ? null : (Entity)this.func_184188_bt().get(0);
+    }
+    //END CM GHAST CHANGES
+
     static class AIFireballAttack extends EntityAIBase
         {
             private final EntityGhast field_179470_b;
@@ -180,7 +230,28 @@
 
             public boolean func_75250_a()
             {
-                return this.field_179470_b.func_70638_az() != null;
+                //return this.parentEntity.getAttackTarget() != null;
+                if (this.field_179470_b.func_70638_az() == null)
+                {
+                    return false;
+                }
+                if (CarpetSettings.rideableGhasts && this.field_179470_b.func_145818_k_() )
+                {
+                    if (this.field_179470_b.func_184207_aI())
+                    {
+                        //reset the attack;
+                        this.func_75251_c();
+                        return false;
+                    }
+                    Entity at = this.field_179470_b.func_70638_az();
+                    if (at instanceof EntityPlayer && GhastHelper.is_yo_bro(this.field_179470_b, (EntityPlayer)at))
+                    {
+                        //reset the attack
+                        this.func_75251_c();
+                        return false;
+                    }
+                }
+                return true;
             }
 
             public void func_75249_e()
@@ -191,6 +262,11 @@
             public void func_75251_c()
             {
                 this.field_179470_b.func_175454_a(false);
+                //CM added reset attacking entity, but this should be here from day one
+                if (CarpetSettings.rideableGhasts)
+                {
+                    this.field_179470_b.func_70624_b(null);
+                }
             }
 
             public void func_75246_d()
