--- ../src-base/minecraft/net/minecraft/entity/monster/EntityGuardian.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityGuardian.java
@@ -36,6 +36,10 @@
 import net.minecraft.world.World;
 import net.minecraft.world.storage.loot.LootTableList;
 
+import net.minecraft.entity.IEntityLivingData;
+import carpet.CarpetSettings;
+import net.minecraft.entity.effect.EntityLightningBolt;
+
 public class EntityGuardian extends EntityMob
 {
     private static final DataParameter<Boolean> field_190766_bz = EntityDataManager.<Boolean>func_187226_a(EntityGuardian.class, DataSerializers.field_187198_h);
@@ -396,6 +400,30 @@
         }
     }
 
+    public void func_70077_a(EntityLightningBolt lightningBolt)
+    {
+        if (!this.field_70170_p.field_72995_K && !this.field_70128_L && CarpetSettings.renewableElderGuardians)
+        {
+            EntityElderGuardian entityguardian = new EntityElderGuardian(this.field_70170_p);
+            entityguardian.func_70012_b(this.field_70165_t, this.field_70163_u, this.field_70161_v, this.field_70177_z, this.field_70125_A);
+            entityguardian.func_180482_a(this.field_70170_p.func_175649_E(new BlockPos(entityguardian)), (IEntityLivingData)null);
+            entityguardian.func_94061_f(this.func_175446_cd());
+
+            if (this.func_145818_k_())
+            {
+                entityguardian.func_96094_a(this.func_95999_t());
+                entityguardian.func_174805_g(this.func_174833_aM());
+            }
+
+            this.field_70170_p.func_72838_d(entityguardian);
+            this.func_70106_y();
+        }
+        else
+        {
+            super.func_70077_a(lightningBolt);
+        }
+    }
+
     static class AIGuardianAttack extends EntityAIBase
         {
             private final EntityGuardian field_179456_a;
