--- ../src-base/minecraft/net/minecraft/block/BlockRedstoneTorch.java
+++ ../src-work/minecraft/net/minecraft/block/BlockRedstoneTorch.java
@@ -19,6 +19,8 @@
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
 
+import carpet.CarpetSettings;
+
 public class BlockRedstoneTorch extends BlockTorch
 {
     private static final Map<World, List<BlockRedstoneTorch.Toggle>> field_150112_b = Maps.<World, List<BlockRedstoneTorch.Toggle>>newHashMap();
@@ -149,7 +151,7 @@
     {
         if (!this.func_176592_e(p_189540_2_, p_189540_3_, p_189540_1_))
         {
-            if (this.field_150113_a == this.func_176597_g(p_189540_2_, p_189540_3_, p_189540_1_))
+            if ( (this.field_150113_a == this.func_176597_g(p_189540_2_, p_189540_3_, p_189540_1_)) && !(CarpetSettings.inconsistentRedstoneTorchesFix && p_189540_2_.func_175691_a(p_189540_3_, this)))
             {
                 p_189540_2_.func_175684_a(p_189540_3_, this, this.func_149738_a(p_189540_2_));
             }
