--- ../src-base/minecraft/net/minecraft/block/BlockBeacon.java
+++ ../src-work/minecraft/net/minecraft/block/BlockBeacon.java
@@ -1,5 +1,6 @@
 package net.minecraft.block;
 
+import carpet.utils.CarpetProfiler;
 import net.minecraft.block.material.MapColor;
 import net.minecraft.block.material.Material;
 import net.minecraft.block.state.IBlockState;
@@ -97,10 +98,17 @@
 
     public static void func_176450_d(final World p_176450_0_, final BlockPos p_176450_1_)
     {
+        if(CarpetProfiler.lastQuestProfilerStart) CarpetProfiler.lastQuestProfiler = true;
         HttpUtil.field_180193_a.submit(new Runnable()
         {
             public void run()
             {
+                if(CarpetProfiler.lastQuestProfiler) {
+                    CarpetProfiler.initThread();
+                    CarpetProfiler.profileLastQuest.get().add(CarpetProfiler.GLASS_START);
+//                    CarpetProfiler.profileLastQuest.get().add(-1000000000L);
+                    CarpetProfiler.profileLastQuest.get().add(System.nanoTime());
+                }
                 Chunk chunk = p_176450_0_.func_175726_f(p_176450_1_);
 
                 for (int i = p_176450_1_.func_177956_o() - 1; i >= 0; --i)
@@ -131,7 +139,16 @@
                         });
                     }
                 }
+                if(CarpetProfiler.lastQuestProfiler) {
+                    CarpetProfiler.profileLastQuest.get().add(CarpetProfiler.GLASS_END);
+                    CarpetProfiler.profileLastQuest.get().add(System.nanoTime());
+                    CarpetProfiler.endLastQuestProfile = false;
+                }
             }
         });
+        if(CarpetProfiler.lastQuestProfiler) {
+            CarpetProfiler.profileLastQuest.get().add(CarpetProfiler.BEACON_FINISH);
+            CarpetProfiler.profileLastQuest.get().add(System.nanoTime());
+        }
     }
 }
