--- ../src-base/minecraft/net/minecraft/block/BlockFenceGate.java
+++ ../src-work/minecraft/net/minecraft/block/BlockFenceGate.java
@@ -20,7 +20,13 @@
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
 
-public class BlockFenceGate extends BlockHorizontal
+import redstone.multimeter.block.MeterableBlock;
+
+import carpet.CarpetSettings;
+
+ import carpet.CarpetSettings;
+
+public class BlockFenceGate extends BlockHorizontal implements MeterableBlock /* RSMM */
 {
     public static final PropertyBool field_176466_a = PropertyBool.func_177716_a("open");
     public static final PropertyBool field_176465_b = PropertyBool.func_177716_a("powered");
@@ -77,7 +83,7 @@
 
     public boolean func_176196_c(World p_176196_1_, BlockPos p_176196_2_)
     {
-        return p_176196_1_.func_180495_p(p_176196_2_.func_177977_b()).func_185904_a().func_76220_a() ? super.func_176196_c(p_176196_1_, p_176196_2_) : false;
+        return (p_176196_1_.func_180495_p(p_176196_2_.func_177977_b()).func_185904_a().func_76220_a()  || CarpetSettings.relaxedBlockPlacement )? super.func_176196_c(p_176196_1_, p_176196_2_) : false; // CM added condition to place in mid air
     }
 
     @Nullable
@@ -144,6 +150,8 @@
         {
             boolean flag = p_189540_2_.func_175640_z(p_189540_3_);
 
+            logPowered(p_189540_2_, p_189540_3_, flag); // RSMM
+
             if (((Boolean)p_189540_1_.func_177229_b(field_176465_b)).booleanValue() != flag)
             {
                 p_189540_2_.func_180501_a(p_189540_3_, p_189540_1_.func_177226_a(field_176465_b, Boolean.valueOf(flag)).func_177226_a(field_176466_a, Boolean.valueOf(flag)), 2);
@@ -195,4 +203,16 @@
             return BlockFaceShape.UNDEFINED;
         }
     }
+
+    // RSMM
+    @Override
+    public boolean logPoweredOnBlockUpdate() {
+        return false;
+    }
+
+    // RSMM
+    @Override
+    public boolean isActive(World world, BlockPos pos, IBlockState state) {
+        return state.func_177229_b(field_176466_a);
+    }
 }
