--- ../src-base/minecraft/net/minecraft/block/BlockStainedGlass.java
+++ ../src-work/minecraft/net/minecraft/block/BlockStainedGlass.java
@@ -14,6 +14,7 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
+import accuratetimer.Test;
 
 public class BlockStainedGlass extends BlockBreakable
 {
@@ -68,6 +69,7 @@
     {
         if (!p_176213_1_.field_72995_K)
         {
+            Test.isGlassRemoved = false;
             BlockBeacon.func_176450_d(p_176213_1_, p_176213_2_);
         }
     }
@@ -76,7 +78,8 @@
     {
         if (!p_180663_1_.field_72995_K)
         {
-            BlockBeacon.func_176450_d(p_180663_1_, p_180663_2_);
+            Test.isGlassRemoved = true;
+//            BlockBeacon.updateColorAsync(worldIn, pos);
         }
     }
 
