--- ../src-base/minecraft/net/minecraft/block/BlockEndPortalFrame.java
+++ ../src-work/minecraft/net/minecraft/block/BlockEndPortalFrame.java
@@ -29,7 +29,9 @@
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
 
-public class BlockEndPortalFrame extends Block
+import redstone.multimeter.block.Meterable;
+
+public class BlockEndPortalFrame extends Block implements Meterable /* RSMM */
 {
     public static final PropertyDirection field_176508_a = BlockHorizontal.field_185512_D;
     public static final PropertyBool field_176507_b = PropertyBool.func_177716_a("eye");
@@ -135,4 +137,10 @@
     {
         return p_193383_4_ == EnumFacing.DOWN ? BlockFaceShape.SOLID : BlockFaceShape.UNDEFINED;
     }
+
+    // RSMM
+    @Override
+    public boolean isActive(World world, BlockPos pos, IBlockState state) {
+        return state.func_177229_b(field_176507_b);
+    }
 }
