--- ../src-base/minecraft/net/minecraft/block/BlockSapling.java
+++ ../src-work/minecraft/net/minecraft/block/BlockSapling.java
@@ -25,6 +25,10 @@
 import net.minecraft.world.gen.feature.WorldGenTrees;
 import net.minecraft.world.gen.feature.WorldGenerator;
 
+import carpet.CarpetSettings;
+import net.minecraft.init.Biomes;
+import carpet.helpers.BlockSaplingHelper;
+
 public class BlockSapling extends BlockBush implements IGrowable
 {
     public static final PropertyEnum<BlockPlanks.EnumType> field_176480_a = PropertyEnum.<BlockPlanks.EnumType>func_177709_a("type", BlockPlanks.EnumType.class);
@@ -68,6 +72,12 @@
         }
         else
         {
+            // Added code for dead shrub rule CARPET-XCOM
+            if(CarpetSettings.desertShrubs && p_176478_1_.func_180494_b(p_176478_2_) == Biomes.field_76769_d && !BlockSaplingHelper.hasWater(p_176478_1_, p_176478_2_))
+            {
+                p_176478_1_.func_180501_a(p_176478_2_, Blocks.field_150330_I.func_176223_P(), 3);
+                return;
+            }
             this.func_176476_e(p_176478_1_, p_176478_2_, p_176478_3_, p_176478_4_);
         }
     }
