--- ../src-base/minecraft/net/minecraft/block/BlockPortal.java
+++ ../src-work/minecraft/net/minecraft/block/BlockPortal.java
@@ -1,5 +1,6 @@
 package net.minecraft.block;
 
+import carpet.utils.portalsearcher.SuperCacheHandler;
 import com.google.common.cache.LoadingCache;
 import java.util.Random;
 import javax.annotation.Nullable;
@@ -24,6 +25,12 @@
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
 
+import carpet.CarpetSettings;
+import net.minecraft.world.WorldServer;
+
+ import net.minecraft.world.WorldServer;
+ import carpet.CarpetSettings;
+
 public class BlockPortal extends BlockBreakable
 {
     public static final PropertyEnum<EnumFacing.Axis> field_176550_a = PropertyEnum.<EnumFacing.Axis>func_177706_a("axis", EnumFacing.Axis.class, EnumFacing.Axis.X, EnumFacing.Axis.Z);
@@ -108,6 +115,12 @@
         if (blockportal$size.func_150860_b() && blockportal$size.field_150864_e == 0)
         {
             blockportal$size.func_150859_c();
+            //CM
+            if (CarpetSettings.portalCaching)
+            {
+                ((WorldServer) p_176548_1_).func_85176_s().clearHistoryCache();
+            }
+            //CM END
             return true;
         }
         else
@@ -117,6 +130,12 @@
             if (blockportal$size1.func_150860_b() && blockportal$size1.field_150864_e == 0)
             {
                 blockportal$size1.func_150859_c();
+                //CM
+                if (CarpetSettings.portalCaching)
+                {
+                    ((WorldServer) p_176548_1_).func_85176_s().clearHistoryCache();
+                }
+                //CM END
                 return true;
             }
             else
@@ -137,6 +156,12 @@
             if (!blockportal$size.func_150860_b() || blockportal$size.field_150864_e < blockportal$size.field_150868_h * blockportal$size.field_150862_g)
             {
                 p_189540_2_.func_175656_a(p_189540_3_, Blocks.field_150350_a.func_176223_P());
+                //CM
+                if (CarpetSettings.portalCaching)
+                {
+                    ((WorldServer) p_189540_2_).func_85176_s().clearHistoryCache();
+                }
+                //CM end
             }
         }
         else if (enumfacing$axis == EnumFacing.Axis.Z)
@@ -146,10 +171,30 @@
             if (!blockportal$size1.func_150860_b() || blockportal$size1.field_150864_e < blockportal$size1.field_150868_h * blockportal$size1.field_150862_g)
             {
                 p_189540_2_.func_175656_a(p_189540_3_, Blocks.field_150350_a.func_176223_P());
+                //CM
+                if (CarpetSettings.portalCaching)
+                {
+                    ((WorldServer) p_189540_2_).func_85176_s().clearHistoryCache();
+                }
+                //CM end
             }
         }
     }
 
+    // betterPortalSearcher
+    @Override
+    public void func_180663_b(World worldIn, BlockPos pos, IBlockState state) {
+        (worldIn.field_73011_w.func_186058_p().func_186068_a() == -1 ? SuperCacheHandler.getHandlerNether() : SuperCacheHandler.getHandlerOverworld()).removePortal(pos);
+        super.func_180663_b(worldIn, pos, state);
+    }
+
+    @Override
+    public void func_176213_c(World worldIn, BlockPos pos, IBlockState state) {
+        (worldIn.field_73011_w.func_186058_p().func_186068_a() == -1 ? SuperCacheHandler.getHandlerNether() : SuperCacheHandler.getHandlerOverworld()).addPortal(pos);
+        super.func_176213_c(worldIn, pos, state);
+    }
+    // end betterPortalSearcher
+
     public int func_149745_a(Random p_149745_1_)
     {
         return 0;
