--- ../src-base/minecraft/net/minecraft/block/BlockFenceGate.java
+++ ../src-work/minecraft/net/minecraft/block/BlockFenceGate.java
@@ -20,6 +20,10 @@
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
 
+import carpet.CarpetSettings;
+
+ import carpet.CarpetSettings;
+
 public class BlockFenceGate extends BlockHorizontal
 {
     public static final PropertyBool field_176466_a = PropertyBool.func_177716_a("open");
@@ -77,7 +81,7 @@
 
     public boolean func_176196_c(World p_176196_1_, BlockPos p_176196_2_)
     {
-        return p_176196_1_.func_180495_p(p_176196_2_.func_177977_b()).func_185904_a().func_76220_a() ? super.func_176196_c(p_176196_1_, p_176196_2_) : false;
+        return (p_176196_1_.func_180495_p(p_176196_2_.func_177977_b()).func_185904_a().func_76220_a()  || CarpetSettings.relaxedBlockPlacement )? super.func_176196_c(p_176196_1_, p_176196_2_) : false; // CM added condition to place in mid air
     }
 
     @Nullable
