--- ../src-base/minecraft/net/minecraft/block/BlockRedstoneRepeater.java
+++ ../src-work/minecraft/net/minecraft/block/BlockRedstoneRepeater.java
@@ -20,6 +20,8 @@
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
 
+import carpet.CarpetSettings;
+
 public class BlockRedstoneRepeater extends BlockRedstoneDiode
 {
     public static final PropertyBool field_176411_a = PropertyBool.func_177716_a("locked");
@@ -64,9 +66,20 @@
         }
     }
 
-    protected int func_176403_d(IBlockState p_176403_1_)
+    protected int getDelay(IBlockState state, World world, BlockPos pos)
     {
-        return ((Integer)p_176403_1_.func_177229_b(field_176410_b)).intValue() * 2;
+    	int delay = 2;
+    	// Added repeater with adjustable delay on terracota CARPET-XCOM
+    	if(CarpetSettings.repeaterPoweredTerracotta){
+            IBlockState iblockstate = world.func_180495_p(pos.func_177977_b());
+            Block block = iblockstate.func_177230_c();
+            if (block == Blocks.field_150406_ce){
+            	delay = block.func_176201_c(iblockstate);
+				if(delay == 0) delay = 100;
+            }
+    	}
+
+        return ((Integer)state.func_177229_b(field_176410_b)).intValue() * delay;
     }
 
     protected IBlockState func_180674_e(IBlockState p_180674_1_)
