--- ../src-base/minecraft/net/minecraft/block/BlockTallGrass.java
+++ ../src-work/minecraft/net/minecraft/block/BlockTallGrass.java
@@ -22,6 +22,8 @@
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
 
+import carpet.helpers.StatHelper;
+
 public class BlockTallGrass extends BlockBush implements IGrowable
 {
     public static final PropertyEnum<BlockTallGrass.EnumType> field_176497_a = PropertyEnum.<BlockTallGrass.EnumType>func_177709_a("type", BlockTallGrass.EnumType.class);
@@ -62,7 +64,10 @@
     {
         if (!p_180657_1_.field_72995_K && p_180657_6_.func_77973_b() == Items.field_151097_aZ)
         {
-            p_180657_2_.func_71029_a(StatList.func_188055_a(this));
+            // CM
+            // player.addStat(StatList.getBlockStats(this));
+            p_180657_2_.func_71029_a(StatHelper.getBlockStateStats(p_180657_4_));
+            // CM END
             func_180635_a(p_180657_1_, p_180657_3_, new ItemStack(Blocks.field_150329_H, 1, ((BlockTallGrass.EnumType)p_180657_4_.func_177229_b(field_176497_a)).func_177044_a()));
         }
         else
