--- ../src-base/minecraft/net/minecraft/block/BlockPumpkin.java
+++ ../src-work/minecraft/net/minecraft/block/BlockPumpkin.java
@@ -26,6 +26,10 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 
+import carpet.CarpetSettings;
+
+ import carpet.CarpetSettings;
+
 public class BlockPumpkin extends BlockHorizontal
 {
     private BlockPattern field_176394_a;
@@ -136,7 +140,7 @@
 
     public boolean func_176196_c(World p_176196_1_, BlockPos p_176196_2_)
     {
-        return p_176196_1_.func_180495_p(p_176196_2_).func_177230_c().field_149764_J.func_76222_j() && p_176196_1_.func_180495_p(p_176196_2_.func_177977_b()).func_185896_q();
+        return p_176196_1_.func_180495_p(p_176196_2_).func_177230_c().field_149764_J.func_76222_j() && ( p_176196_1_.func_180495_p(p_176196_2_.func_177977_b()).func_185896_q() || CarpetSettings.relaxedBlockPlacement); // CM added condition for easier placement
     }
 
     public IBlockState func_185499_a(IBlockState p_185499_1_, Rotation p_185499_2_)
