--- ../src-base/minecraft/net/minecraft/block/BlockFurnace.java
+++ ../src-work/minecraft/net/minecraft/block/BlockFurnace.java
@@ -24,6 +24,8 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 
+import carpet.CarpetSettings;
+
 public class BlockFurnace extends BlockContainer
 {
     public static final PropertyDirection field_176447_a = BlockHorizontal.field_185512_D;
@@ -44,6 +46,8 @@
 
     public void func_176213_c(World p_176213_1_, BlockPos p_176213_2_, IBlockState p_176213_3_)
     {
+    	// A bug that causes bad rotations fixed in 1.13 CARPET-XCOM #not needed in 1.13
+    	if(CarpetSettings.accurateBlockPlacement) { return; }
         this.func_176445_e(p_176213_1_, p_176213_2_, p_176213_3_);
     }
 
