--- ../src-base/minecraft/net/minecraft/block/BlockDynamicLiquid.java
+++ ../src-work/minecraft/net/minecraft/block/BlockDynamicLiquid.java
@@ -3,11 +3,14 @@
 import java.util.EnumSet;
 import java.util.Random;
 import java.util.Set;
+
+import carpet.CarpetSettings;
 import net.minecraft.block.material.Material;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.init.Blocks;
 import net.minecraft.util.EnumFacing;
 import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.World;
 
 public class BlockDynamicLiquid extends BlockLiquid
@@ -134,7 +137,7 @@
         }
         else if (i >= 0 && (i == 0 || this.func_176372_g(p_180650_1_, p_180650_2_.func_177977_b(), iblockstate1)))
         {
-            Set<EnumFacing> set = this.func_176376_e(p_180650_1_, p_180650_2_);
+            Set<EnumFacing> set = this.getPossibleFlowDirections(p_180650_1_, p_180650_2_, i);
             int k1 = i + j;
 
             if (i >= 8)
@@ -213,23 +216,31 @@
         return this.field_149764_J == Material.field_151587_i && !p_185698_1_.field_73011_w.func_177500_n() ? 2 : 4;
     }
 
-    private Set<EnumFacing> func_176376_e(World p_176376_1_, BlockPos p_176376_2_)
+    private Set<EnumFacing> getPossibleFlowDirections(World worldIn, BlockPos pos, int level)
     {
         int i = 1000;
         Set<EnumFacing> set = EnumSet.<EnumFacing>noneOf(EnumFacing.class);
 
+        int maxflow = 1;
+        if (CarpetSettings.waterFlow != CarpetSettings.WaterFlow.vanilla && level < 8)
+        {
+            maxflow = level - (CarpetSettings.waterFlow == CarpetSettings.WaterFlow.correct ? 1 : 3); // level = 7 is the shallowest water
+            if (maxflow < 1) maxflow = 1;
+        }
+
+
         for (EnumFacing enumfacing : EnumFacing.Plane.HORIZONTAL)
         {
-            BlockPos blockpos = p_176376_2_.func_177972_a(enumfacing);
-            IBlockState iblockstate = p_176376_1_.func_180495_p(blockpos);
+            BlockPos blockpos = pos.func_177972_a(enumfacing);
+            IBlockState iblockstate = worldIn.func_180495_p(blockpos);
 
-            if (!this.func_176372_g(p_176376_1_, blockpos, iblockstate) && (iblockstate.func_185904_a() != this.field_149764_J || ((Integer)iblockstate.func_177229_b(field_176367_b)).intValue() > 0))
+            if (!this.func_176372_g(worldIn, blockpos, iblockstate) && (iblockstate.func_185904_a() != this.field_149764_J || ((Integer)iblockstate.func_177229_b(field_176367_b)).intValue() > 0))
             {
                 int j;
 
-                if (this.func_176372_g(p_176376_1_, blockpos.func_177977_b(), p_176376_1_.func_180495_p(blockpos.func_177977_b())))
+                if (this.func_176372_g(worldIn, blockpos.func_177977_b(), worldIn.func_180495_p(blockpos.func_177977_b())))
                 {
-                    j = this.func_176374_a(p_176376_1_, blockpos, 1, enumfacing.func_176734_d());
+                    j = this.func_176374_a(worldIn, blockpos, maxflow, enumfacing.func_176734_d()); // maxflow = 1
                 }
                 else
                 {
