--- ../src-base/minecraft/net/minecraft/block/BlockPistonMoving.java
+++ ../src-work/minecraft/net/minecraft/block/BlockPistonMoving.java
@@ -3,6 +3,8 @@
 import java.util.List;
 import java.util.Random;
 import javax.annotation.Nullable;
+
+import carpet.CarpetSettings;
 import net.minecraft.block.material.Material;
 import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyDirection;
@@ -46,9 +48,12 @@
         return null;
     }
 
-    public static TileEntity func_185588_a(IBlockState p_185588_0_, EnumFacing p_185588_1_, boolean p_185588_2_, boolean p_185588_3_)
+    public static TileEntity createTilePiston(IBlockState blockStateIn, EnumFacing facingIn, boolean extendingIn, boolean shouldHeadBeRenderedIn, int lightOpacity, int lightValue)
     {
-        return new TileEntityPiston(p_185588_0_, p_185588_1_, p_185588_2_, p_185588_3_);
+        TileEntityPiston te = new TileEntityPiston(blockStateIn, facingIn, extendingIn, shouldHeadBeRenderedIn);
+        te.lightOpactiy = lightOpacity;
+        te.lightValue = lightValue;
+        return te;
     }
 
     public void func_180663_b(World p_180663_1_, BlockPos p_180663_2_, IBlockState p_180663_3_)
@@ -214,4 +219,39 @@
     {
         return BlockFaceShape.UNDEFINED;
     }
+
+    // CM: movingBlockLightOptimization
+    @Override
+    public int getLightOpacity(IBlockAccess world, BlockPos pos, IBlockState state)
+    {
+        if (!CarpetSettings.movingBlockLightOptimization)
+        {
+            return super.func_149717_k(state);
+        }
+
+        TileEntity te = world.func_175625_s(pos);
+        if (!(te instanceof TileEntityPiston))
+        {
+            return super.func_149717_k(state);
+        }
+
+        return ((TileEntityPiston) te).lightOpactiy;
+    }
+
+    @Override
+    public int getLightValue(IBlockAccess world, BlockPos pos, IBlockState state)
+    {
+        if (!CarpetSettings.movingBlockLightOptimization)
+        {
+            return super.func_149750_m(state);
+        }
+
+        TileEntity te = world.func_175625_s(pos);
+        if (!(te instanceof TileEntityPiston))
+        {
+            return super.func_149750_m(state);
+        }
+
+        return ((TileEntityPiston) te).lightValue;
+    }
 }
