--- ../src-base/minecraft/net/minecraft/block/Block.java
+++ ../src-work/minecraft/net/minecraft/block/Block.java
@@ -1,11 +1,15 @@
 package net.minecraft.block;
 
+import carpet.CarpetServer;
+import carpet.CarpetSettings;
 import com.google.common.collect.Sets;
 import com.google.common.collect.UnmodifiableIterator;
 import java.util.List;
 import java.util.Random;
 import java.util.Set;
 import javax.annotation.Nullable;
+
+import net.minecraft.advancements.critereon.PlacedBlockTrigger;
 import net.minecraft.block.material.EnumPushReaction;
 import net.minecraft.block.material.MapColor;
 import net.minecraft.block.material.Material;
@@ -23,20 +27,16 @@
 import net.minecraft.init.Blocks;
 import net.minecraft.init.Enchantments;
 import net.minecraft.init.Items;
+import net.minecraft.init.SoundEvents;
 import net.minecraft.item.EnumDyeColor;
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemBlock;
 import net.minecraft.item.ItemStack;
+import net.minecraft.network.play.server.SPacketSoundEffect;
+import net.minecraft.server.management.PlayerInteractionManager;
 import net.minecraft.stats.StatList;
 import net.minecraft.tileentity.TileEntity;
-import net.minecraft.util.EnumBlockRenderType;
-import net.minecraft.util.EnumFacing;
-import net.minecraft.util.EnumHand;
-import net.minecraft.util.Mirror;
-import net.minecraft.util.NonNullList;
-import net.minecraft.util.ObjectIntIdentityMap;
-import net.minecraft.util.ResourceLocation;
-import net.minecraft.util.Rotation;
+import net.minecraft.util.*;
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.MathHelper;
@@ -48,6 +48,10 @@
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
 
+import carpet.helpers.CapturedDrops;
+import carpet.helpers.StatHelper;
+import carpet.patches.BlockWool;
+
 public class Block
 {
     private static final ResourceLocation field_176230_a = new ResourceLocation("air");
@@ -74,7 +78,7 @@
     public float field_149765_K;
     protected final BlockStateContainer field_176227_L;
     private IBlockState field_176228_M;
-    private String field_149770_b;
+    public String field_149770_b;
 
     public static int func_149682_b(Block p_149682_0_)
     {
@@ -236,7 +240,7 @@
         return this;
     }
 
-    protected Block func_149713_g(int p_149713_1_)
+    public Block func_149713_g(int p_149713_1_)
     {
         this.field_149786_r = p_149713_1_;
         return this;
@@ -328,7 +332,7 @@
         return this.field_149782_v;
     }
 
-    protected Block func_149675_a(boolean p_149675_1_)
+    public Block func_149675_a(boolean p_149675_1_)
     {
         this.field_149789_z = p_149675_1_;
         return this;
@@ -400,7 +404,10 @@
 
     public void func_180645_a(World p_180645_1_, BlockPos p_180645_2_, IBlockState p_180645_3_, Random p_180645_4_)
     {
+        if(carpet.carpetclient.CarpetClientChunkLogger.logger.enabled)
+            carpet.carpetclient.CarpetClientChunkLogger.setReason("Randomtick block: " + p_180645_3_.func_177230_c().func_149732_F());
         this.func_180650_b(p_180645_1_, p_180645_2_, p_180645_3_, p_180645_4_);
+        carpet.carpetclient.CarpetClientChunkLogger.resetReason();
     }
 
     public void func_180650_b(World p_180650_1_, BlockPos p_180650_2_, IBlockState p_180650_3_, Random p_180650_4_)
@@ -489,8 +496,17 @@
             double d1 = (double)(p_180635_0_.field_73012_v.nextFloat() * 0.5F) + 0.25D;
             double d2 = (double)(p_180635_0_.field_73012_v.nextFloat() * 0.5F) + 0.25D;
             EntityItem entityitem = new EntityItem(p_180635_0_, (double)p_180635_1_.func_177958_n() + d0, (double)p_180635_1_.func_177956_o() + d1, (double)p_180635_1_.func_177952_p() + d2, p_180635_2_);
+            if(CarpetSettings.carefulBreak && PlayerInteractionManager.playerMinedBlock != null && PlayerInteractionManager.playerMinedBlock.func_70093_af()){
+                entityitem.func_70100_b_(PlayerInteractionManager.playerMinedBlock);
+                if(entityitem.field_70128_L){
+                    PlayerInteractionManager.playerMinedBlock.field_71135_a.func_147359_a(new SPacketSoundEffect(SoundEvents.field_187638_cR, SoundCategory.BLOCKS, p_180635_1_.func_177958_n(), p_180635_1_.func_177956_o(), p_180635_1_.func_177952_p(), 1.0F, (CarpetServer.rand.nextFloat() - CarpetServer.rand.nextFloat()) * 1.4F + 2.0F));
+                    return;
+                }
+            }
             entityitem.func_174869_p();
             p_180635_0_.func_72838_d(entityitem);
+            if (CapturedDrops.isCapturingDrops())
+                CapturedDrops.captureDrop(entityitem);
         }
     }
 
@@ -594,7 +610,10 @@
 
     public void func_180657_a(World p_180657_1_, EntityPlayer p_180657_2_, BlockPos p_180657_3_, IBlockState p_180657_4_, @Nullable TileEntity p_180657_5_, ItemStack p_180657_6_)
     {
-        p_180657_2_.func_71029_a(StatList.func_188055_a(this));
+        // CM
+        // player.addStat(StatList.getBlockStats(this));
+        p_180657_2_.func_71029_a(StatHelper.getBlockStateStats(p_180657_4_));
+        // CM END
         p_180657_2_.func_71020_j(0.005F);
 
         if (this.func_149700_E() && EnchantmentHelper.func_77506_a(Enchantments.field_185306_r, p_180657_6_) > 0)
@@ -849,7 +868,8 @@
         func_176219_a(32, "deadbush", (new BlockDeadBush()).func_149711_c(0.0F).func_149672_a(SoundType.field_185850_c).func_149663_c("deadbush"));
         func_176219_a(33, "piston", (new BlockPistonBase(false)).func_149663_c("pistonBase"));
         func_176219_a(34, "piston_head", (new BlockPistonExtension()).func_149663_c("pistonBase"));
-        func_176219_a(35, "wool", (new BlockColored(Material.field_151580_n)).func_149711_c(0.8F).func_149672_a(SoundType.field_185854_g).func_149663_c("cloth"));
+        // CM: use custom wool block
+        func_176219_a(35, "wool", (new BlockWool()).func_149711_c(0.8F).func_149672_a(SoundType.field_185854_g).func_149663_c("cloth"));
         func_176219_a(36, "piston_extension", new BlockPistonMoving());
         func_176219_a(37, "yellow_flower", (new BlockYellowFlower()).func_149711_c(0.0F).func_149672_a(SoundType.field_185850_c).func_149663_c("flower1"));
         func_176219_a(38, "red_flower", (new BlockRedFlower()).func_149711_c(0.0F).func_149672_a(SoundType.field_185850_c).func_149663_c("flower2"));
