--- ../src-base/minecraft/net/minecraft/block/BlockCarpet.java
+++ ../src-work/minecraft/net/minecraft/block/BlockCarpet.java
@@ -1,5 +1,6 @@
 package net.minecraft.block;
 
+import carpet.CarpetSettings;
 import net.minecraft.block.material.MapColor;
 import net.minecraft.block.material.Material;
 import net.minecraft.block.properties.IProperty;
@@ -17,6 +18,10 @@
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
 
+import net.minecraft.entity.EntityLivingBase;
+import carpet.utils.WoolTool;
+import net.minecraft.entity.player.EntityPlayer;
+
 public class BlockCarpet extends Block
 {
     public static final PropertyEnum<EnumDyeColor> field_176330_a = PropertyEnum.<EnumDyeColor>func_177709_a("color", EnumDyeColor.class);
@@ -57,6 +62,7 @@
 
     public void func_189540_a(IBlockState p_189540_1_, World p_189540_2_, BlockPos p_189540_3_, Block p_189540_4_, BlockPos p_189540_5_)
     {
+        if(CarpetSettings.duplicationFixMovingCarpets && BlockPistonBase.isBeingPushed(p_189540_3_)) return; // Added duplication fix for moving carpets CARPET-XCOM
         this.func_176328_e(p_189540_2_, p_189540_3_, p_189540_1_);
     }
 
@@ -111,4 +117,13 @@
     {
         return p_193383_4_ == EnumFacing.DOWN ? BlockFaceShape.SOLID : BlockFaceShape.UNDEFINED;
     }
+    public IBlockState func_180642_a(World worldIn, BlockPos pos, EnumFacing facing, float hitX, float hitY, float hitZ, int meta, EntityLivingBase placer)
+    {
+        IBlockState state = super.func_180642_a(worldIn, pos, facing, hitX, hitY, hitZ, meta, placer);
+        if (placer instanceof EntityPlayer)
+        {
+             WoolTool.carpetPlacedAction(((EnumDyeColor)state.func_177229_b(field_176330_a)), (EntityPlayer)placer, pos, worldIn);
+        }
+        return state;
+    }
 }
