--- ../src-base/minecraft/net/minecraft/block/BlockNote.java
+++ ../src-work/minecraft/net/minecraft/block/BlockNote.java
@@ -19,9 +19,15 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 
-public class BlockNote extends BlockContainer
+import redstone.multimeter.block.MeterableBlock;
+import redstone.multimeter.helper.WorldHelper;
+import carpet.CarpetSettings;
+import net.minecraft.init.Blocks;
+
+public class BlockNote extends BlockContainer implements MeterableBlock /* RSMM */
 {
     private static final List<SoundEvent> field_176434_a = Lists.newArrayList(SoundEvents.field_187682_dG, SoundEvents.field_187676_dE, SoundEvents.field_187688_dI, SoundEvents.field_187685_dH, SoundEvents.field_187679_dF, SoundEvents.field_193809_ey, SoundEvents.field_193807_ew, SoundEvents.field_193810_ez, SoundEvents.field_193808_ex, SoundEvents.field_193785_eE);
+    private SoundEvent saveSoundEvent;
 
     public BlockNote()
     {
@@ -34,9 +40,13 @@
         boolean flag = p_189540_2_.func_175640_z(p_189540_3_);
         TileEntity tileentity = p_189540_2_.func_175625_s(p_189540_3_);
 
+        logPowered(p_189540_2_, p_189540_3_, flag); // RSMM
+
         if (tileentity instanceof TileEntityNote)
         {
             TileEntityNote tileentitynote = (TileEntityNote)tileentity;
+            // Get the instrument type based on the block below. CARPET-XCOM
+            SoundEvent instrument = func_185576_e(getID(p_189540_2_, p_189540_3_));
 
             if (tileentitynote.field_145880_i != flag)
             {
@@ -46,6 +56,18 @@
                 }
 
                 tileentitynote.field_145880_i = flag;
+  
+                // RSMM start
+                if (!p_189540_2_.field_72995_K) {
+                    WorldHelper.getMultimeter().logActive(p_189540_2_, p_189540_3_, flag);
+                }
+                // RSMM end
+
+                //Added note block imitation in 1.13 CARPET-XCOM
+                if(CarpetSettings.noteBlockImitationOf1_13) p_189540_2_.func_175685_c(p_189540_3_, this, true);
+            }else if(saveSoundEvent == null || saveSoundEvent != instrument){
+	            saveSoundEvent = instrument;
+	            if(CarpetSettings.noteBlockImitationOf1_13) p_189540_2_.func_190522_c(p_189540_3_, p_189540_4_);
             }
         }
     }
@@ -68,6 +90,9 @@
                 p_180639_4_.func_71029_a(StatList.field_188087_U);
             }
 
+            //Added note block imitation in 1.13 CARPET-XCOM
+            if(CarpetSettings.noteBlockImitationOf1_13) p_180639_1_.func_175685_c(p_180639_2_, this, true);
+
             return true;
         }
     }
@@ -113,4 +138,78 @@
     {
         return EnumBlockRenderType.MODEL;
     }
+
+    // Getting block id to store value of note block sound type. CARPET-XCOM
+    private int getID(World worldIn, BlockPos pos){
+        IBlockState iblockstate = worldIn.func_180495_p(pos.func_177977_b());
+        Material material = iblockstate.func_185904_a();
+        int i = 0;
+
+        if (material == Material.field_151576_e)
+        {
+            i = 1;
+        }
+
+        if (material == Material.field_151595_p)
+        {
+            i = 2;
+        }
+
+        if (material == Material.field_151592_s)
+        {
+            i = 3;
+        }
+
+        if (material == Material.field_151575_d)
+        {
+            i = 4;
+        }
+
+        Block block = iblockstate.func_177230_c();
+
+        if (block == Blocks.field_150435_aG)
+        {
+            i = 5;
+        }
+
+        if (block == Blocks.field_150340_R)
+        {
+            i = 6;
+        }
+
+        if (block == Blocks.field_150325_L)
+        {
+            i = 7;
+        }
+
+        if (block == Blocks.field_150403_cj)
+        {
+            i = 8;
+        }
+
+        if (block == Blocks.field_189880_di)
+        {
+            i = 9;
+        }
+
+        return i;
+    }
+
+    // RSMM
+    @Override
+    public boolean logPoweredOnBlockUpdate() {
+        return false;
+    }
+
+    // RSMM
+    @Override
+    public boolean isActive(World world, BlockPos pos, IBlockState state) {
+        TileEntity blockEntity = world.func_175625_s(pos);
+
+        if (blockEntity instanceof TileEntityNote) {
+            return ((TileEntityNote)blockEntity).field_145880_i;
+        }
+
+        return false;
+    }
 }
