--- ../src-base/minecraft/net/minecraft/block/BlockDispenser.java
+++ ../src-work/minecraft/net/minecraft/block/BlockDispenser.java
@@ -32,7 +32,11 @@
 import net.minecraft.util.registry.RegistryDefaulted;
 import net.minecraft.world.World;
 
-public class BlockDispenser extends BlockContainer
+import redstone.multimeter.block.MeterableBlock;
+
+import carpet.CarpetSettings;
+
+public class BlockDispenser extends BlockContainer implements MeterableBlock /* RSMM */
 {
     public static final PropertyDirection field_176441_a = BlockDirectional.field_176387_N;
     public static final PropertyBool field_176440_b = PropertyBool.func_177716_a("triggered");
@@ -54,6 +58,8 @@
     public void func_176213_c(World p_176213_1_, BlockPos p_176213_2_, IBlockState p_176213_3_)
     {
         super.func_176213_c(p_176213_1_, p_176213_2_, p_176213_3_);
+        // A bug that causes bad rotations fixed in 1.13 CARPET-XCOM -> not needed in 1.13
+        if(CarpetSettings.accurateBlockPlacement) { return; }
         this.func_176438_e(p_176213_1_, p_176213_2_, p_176213_3_);
     }
 
@@ -153,9 +159,21 @@
 
     public void func_189540_a(IBlockState p_189540_1_, World p_189540_2_, BlockPos p_189540_3_, Block p_189540_4_, BlockPos p_189540_5_)
     {
-        boolean flag = p_189540_2_.func_175640_z(p_189540_3_) || p_189540_2_.func_175640_z(p_189540_3_.func_177984_a());
+        /* carpet mod instead of boolean flag = worldIn.isBlockPowered(pos) || worldIn.isBlockPowered(pos.up()); */
+        boolean flag = false;
+        if (CarpetSettings.quasiConnectivity)
+        {
+            flag = p_189540_2_.func_175640_z(p_189540_3_) || p_189540_2_.func_175640_z(p_189540_3_.func_177984_a());
+        }
+        else
+        {
+            flag = p_189540_2_.func_175640_z(p_189540_3_);
+        }
+        /* end */
         boolean flag1 = ((Boolean)p_189540_1_.func_177229_b(field_176440_b)).booleanValue();
 
+        logPowered(p_189540_2_, p_189540_3_, flag); // RSMM
+
         if (flag && !flag1)
         {
             p_189540_2_.func_175684_a(p_189540_3_, this, this.func_149738_a(p_189540_2_));
@@ -187,7 +205,8 @@
 
     public void func_180633_a(World p_180633_1_, BlockPos p_180633_2_, IBlockState p_180633_3_, EntityLivingBase p_180633_4_, ItemStack p_180633_5_)
     {
-        p_180633_1_.func_180501_a(p_180633_2_, p_180633_3_.func_177226_a(field_176441_a, EnumFacing.func_190914_a(p_180633_2_, p_180633_4_)), 2);
+        if (!CarpetSettings.accurateBlockPlacement)
+            p_180633_1_.func_180501_a(p_180633_2_, p_180633_3_.func_177226_a(field_176441_a, EnumFacing.func_190914_a(p_180633_2_, p_180633_4_)), 2);
 
         if (p_180633_5_.func_82837_s())
         {
@@ -269,4 +288,22 @@
     {
         return new BlockStateContainer(this, new IProperty[] {field_176441_a, field_176440_b});
     }
+
+    // RSMM
+    @Override
+    public boolean logPoweredOnBlockUpdate() {
+        return false;
+    }
+
+    // RSMM
+    @Override
+    public boolean isPowered(World world, BlockPos pos, IBlockState state) {
+        return world.func_175640_z(pos) || (CarpetSettings.quasiConnectivity && world.func_175640_z(pos.func_177984_a()));
+    }
+
+    // RSMM
+    @Override
+    public boolean isActive(World world, BlockPos pos, IBlockState state) {
+        return state.func_177229_b(field_176440_b);
+    }
 }
