--- ../src-base/minecraft/net/minecraft/block/Block.java
+++ ../src-work/minecraft/net/minecraft/block/Block.java
@@ -47,6 +47,8 @@
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
 
+import carpet.helpers.CapturedDrops;
+
 public class Block
 {
     private static final ResourceLocation field_176230_a = new ResourceLocation("air");
@@ -73,7 +75,7 @@
     public float field_149765_K;
     protected final BlockStateContainer field_176227_L;
     private IBlockState field_176228_M;
-    private String field_149770_b;
+    public String field_149770_b;
 
     public static int func_149682_b(Block p_149682_0_)
     {
@@ -235,7 +237,7 @@
         return this;
     }
 
-    protected Block func_149713_g(int p_149713_1_)
+    public Block func_149713_g(int p_149713_1_)
     {
         this.field_149786_r = p_149713_1_;
         return this;
@@ -327,7 +329,7 @@
         return this.field_149782_v;
     }
 
-    protected Block func_149675_a(boolean p_149675_1_)
+    public Block func_149675_a(boolean p_149675_1_)
     {
         this.field_149789_z = p_149675_1_;
         return this;
@@ -399,7 +401,10 @@
 
     public void func_180645_a(World p_180645_1_, BlockPos p_180645_2_, IBlockState p_180645_3_, Random p_180645_4_)
     {
+        if(carpet.carpetclient.CarpetClientChunkLogger.logger.enabled)
+            carpet.carpetclient.CarpetClientChunkLogger.setReason("Randomtick block: " + p_180645_3_.func_177230_c().func_149732_F());
         this.func_180650_b(p_180645_1_, p_180645_2_, p_180645_3_, p_180645_4_);
+        carpet.carpetclient.CarpetClientChunkLogger.resetReason();
     }
 
     public void func_180650_b(World p_180650_1_, BlockPos p_180650_2_, IBlockState p_180650_3_, Random p_180650_4_)
@@ -490,6 +495,8 @@
             EntityItem entityitem = new EntityItem(p_180635_0_, (double)p_180635_1_.func_177958_n() + d0, (double)p_180635_1_.func_177956_o() + d1, (double)p_180635_1_.func_177952_p() + d2, p_180635_2_);
             entityitem.func_174869_p();
             p_180635_0_.func_72838_d(entityitem);
+            if (CapturedDrops.isCapturingDrops())
+                CapturedDrops.captureDrop(entityitem);
         }
     }
 
