--- ../src-base/minecraft/net/minecraft/command/server/CommandSetBlock.java
+++ ../src-work/minecraft/net/minecraft/command/server/CommandSetBlock.java
@@ -20,6 +20,10 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 
+import carpet.CarpetSettings;
+
+ import carpet.CarpetSettings;
+
 public class CommandSetBlock extends CommandBase
 {
     public String func_71517_b()
@@ -110,7 +114,7 @@
                     ((IInventory)tileentity1).func_174888_l();
                 }
 
-                if (!world.func_180501_a(blockpos, iblockstate, 2))
+                if (!world.func_180501_a(blockpos, iblockstate, 2 | (CarpetSettings.getBool("fillUpdates")?0:128) )) // CM
                 {
                     throw new CommandException("commands.setblock.noChange", new Object[0]);
                 }
@@ -129,7 +133,12 @@
                         }
                     }
 
-                    world.func_175722_b(blockpos, iblockstate.func_177230_c(), false);
+                    // CM
+                    if (CarpetSettings.getBool("fillUpdates"))
+                    {
+                        world.func_175722_b(blockpos, iblockstate.func_177230_c(), false);
+                    }
+                    // CM end
                     p_184881_2_.func_174794_a(CommandResultStats.Type.AFFECTED_BLOCKS, 1);
                     func_152373_a(p_184881_2_, this, "commands.setblock.success", new Object[0]);
                 }
