--- ../src-base/minecraft/net/minecraft/item/crafting/CraftingManager.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/CraftingManager.java
@@ -32,6 +32,8 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import carpet.helpers.FireworksCraftingRecipes;
+
 public class CraftingManager
 {
     private static final Logger field_192422_a = LogManager.getLogger();
@@ -42,6 +44,14 @@
     {
         try
         {
+        	// Added rockets to the list of crafting recipes CARPET-XCOM
+            // will add later if found a way to hide popups for vanilla clients
+        	FireworksCraftingRecipes.createRockets();
+        	func_193379_a("rocket1", FireworksCraftingRecipes.rocketOne);
+        	func_193379_a("rocket2", FireworksCraftingRecipes.rocketTwo);
+        	func_193379_a("rocket3", FireworksCraftingRecipes.rocketThree);
+        	// End
+        	
             func_193379_a("armordye", new RecipesArmorDyes());
             func_193379_a("bookcloning", new RecipeBookCloning());
             func_193379_a("mapcloning", new RecipesMapCloning());
