--- ../src-base/minecraft/net/minecraft/item/crafting/CraftingManager.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/CraftingManager.java
@@ -32,9 +32,13 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import java.util.ArrayList;
+import java.util.List;
+import carpet.helpers.CustomCrafting;
+
 public class CraftingManager
 {
-    private static final Logger field_192422_a = LogManager.getLogger();
+    public static final Logger field_192422_a = LogManager.getLogger();
     private static int field_193381_c;
     public static final RegistryNamespaced<ResourceLocation, IRecipe> field_193380_a = new RegistryNamespaced<ResourceLocation, IRecipe>();
 
@@ -53,7 +57,9 @@
             func_193379_a("banneraddpattern", new RecipesBanners.RecipeAddPattern());
             func_193379_a("shielddecoration", new ShieldRecipes.Decoration());
             func_193379_a("shulkerboxcoloring", new ShulkerBoxRecipes.ShulkerBoxColoring());
-            return func_192420_c();
+            boolean result = func_192420_c();
+            result = CustomCrafting.registerCustomRecipes(result);
+            return result;
         }
         catch (Throwable var1)
         {
@@ -95,6 +101,16 @@
 
                 Iterator<Path> iterator = Files.walk(path).iterator();
 
+                // Hardcode the correctly sorted recipes to fix instances when users scrambles this list. CARPET-XCOM
+                List<Path> list = new ArrayList<>();
+                while (iterator.hasNext()) {
+                    Path path1 = iterator.next();
+                    list.add(path1);
+                }
+                Collections.sort(list, (a, b) -> b.toString().compareTo(a.toString()));
+                iterator = list.iterator();
+                //
+
                 while (iterator.hasNext())
                 {
                     Path path1 = iterator.next();
@@ -155,7 +171,7 @@
         return flag1;
     }
 
-    private static IRecipe func_193376_a(JsonObject p_193376_0_)
+    public static IRecipe func_193376_a(JsonObject p_193376_0_)
     {
         String s = JsonUtils.func_151200_h(p_193376_0_, "type");
 
