--- ../src-base/minecraft/net/minecraft/item/crafting/ShapedRecipes.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/ShapedRecipes.java
@@ -59,11 +59,6 @@
         return nonnulllist;
     }
 
-    public NonNullList<Ingredient> func_192400_c()
-    {
-        return this.field_77574_d;
-    }
-
     public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
     {
         for (int i = 0; i <= 3 - this.field_77576_b; ++i)
@@ -365,4 +360,7 @@
             return new ItemStack(item, j, i);
         }
     }
+
+    // Client code required for nitwit crafting
+    public NonNullList<Ingredient> func_192400_c() { return this.field_77574_d; }
 }
