--- ../src-base/minecraft/net/minecraft/item/crafting/ShapelessRecipes.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/ShapelessRecipes.java
@@ -29,11 +29,6 @@
         return this.field_77580_a;
     }
 
-    public NonNullList<Ingredient> func_192400_c()
-    {
-        return this.field_77579_b;
-    }
-
     public NonNullList<ItemStack> func_179532_b(InventoryCrafting p_179532_1_)
     {
         NonNullList<ItemStack> nonnulllist = NonNullList.<ItemStack>func_191197_a(p_179532_1_.func_70302_i_(), ItemStack.field_190927_a);
@@ -127,4 +122,7 @@
 
         return nonnulllist;
     }
+
+    // Client code required for nitwit crafting
+    public NonNullList<Ingredient> func_192400_c() { return this.field_77579_b; }
 }
