--- ../src-base/minecraft/net/minecraft/util/ServerRecipeBookHelper.java
+++ ../src-work/minecraft/net/minecraft/util/ServerRecipeBookHelper.java
@@ -1,5 +1,8 @@
 package net.minecraft.util;
 
+import carpet.CarpetSettings;
+import carpet.logging.LoggerRegistry;
+import carpet.utils.Messenger;
 import it.unimi.dsi.fastutil.ints.IntArrayList;
 import it.unimi.dsi.fastutil.ints.IntList;
 import it.unimi.dsi.fastutil.ints.IntListIterator;
@@ -7,6 +10,7 @@
 import java.util.List;
 import javax.annotation.Nullable;
 import net.minecraft.client.util.RecipeItemHelper;
+import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.entity.player.InventoryPlayer;
 import net.minecraft.inventory.Container;
@@ -16,9 +20,11 @@
 import net.minecraft.inventory.InventoryCrafting;
 import net.minecraft.inventory.Slot;
 import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.CraftingManager;
 import net.minecraft.item.crafting.IRecipe;
 import net.minecraft.item.crafting.ShapedRecipes;
 import net.minecraft.network.play.server.SPacketPlaceGhostRecipe;
+import net.minecraft.util.text.ITextComponent;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -35,6 +41,18 @@
 
     public void func_194327_a(EntityPlayerMP p_194327_1_, @Nullable IRecipe p_194327_2_, boolean p_194327_3_)
     {
+
+        if(LoggerRegistry.__recipes){ // Added debugger for the instance people need help debuging why there recipes don't work. CARPET-XCOM
+            int i = 0;
+            for(ResourceLocation r : CraftingManager.field_193380_a.func_148742_b()) {
+                String index = Integer.toString(i);
+                LoggerRegistry.getLogger("recipes").log(()-> new ITextComponent[]{
+                        Messenger.s(null, index + ": " + r.func_110623_a())
+                });
+                i++;
+            }
+        }
+
         if (p_194327_2_ != null && p_194327_1_.func_192037_E().func_193830_f(p_194327_2_))
         {
             this.field_194332_c = p_194327_1_;
@@ -48,7 +66,8 @@
             if (container instanceof ContainerWorkbench)
             {
                 this.field_194335_f = ((ContainerWorkbench)container).field_75160_f;
-                this.field_194336_g = ((ContainerWorkbench)container).field_75162_e;
+                // Method added for auto crafting CARPET-XCOM
+                this.field_194336_g = ((ContainerWorkbench)container).getInventoryCrafting();
             }
             else if (container instanceof ContainerPlayer)
             {
@@ -67,6 +86,8 @@
                     if (this.field_194331_b.func_194116_a(p_194327_2_, (IntList)null))
                     {
                         this.func_194329_b();
+                        // Intentional duping bug added back for compatibility with 12.0, community request. CARPET-XCOM
+                        if(CarpetSettings.craftingWindowDuplication) craftingWindowDupingBugAddedBack(p_194327_1_);
                     }
                     else
                     {
@@ -304,4 +325,29 @@
 
         return true;
     }
+
+    // Intentional duping bug added back for compatibility with 12.0, community request. CARPET-XCOM
+    private void craftingWindowDupingBugAddedBack(EntityPlayerMP player){
+        int slot = player.getDupeItem();
+        if(slot == Integer.MIN_VALUE) return;
+        ItemStack dupeItem = player.field_71071_by.func_70301_a(slot);
+        if(dupeItem.func_190926_b()) return;
+
+        int size = dupeItem.func_190916_E();
+
+        for (int j = 0; j < this.field_194336_g.func_70302_i_(); ++j)
+        {
+            ItemStack itemstack = this.field_194336_g.func_70301_a(j);
+
+            if (!itemstack.func_190926_b())
+            {
+                size += itemstack.func_190916_E();
+                itemstack.func_190920_e(0);
+            }
+        }
+
+        dupeItem.forceStackSize(size);
+        field_194335_f.func_174888_l();
+        player.clearDupeItem();
+    }
 }
