--- ../src-base/minecraft/net/minecraft/util/ServerRecipeBookHelper.java
+++ ../src-work/minecraft/net/minecraft/util/ServerRecipeBookHelper.java
@@ -22,6 +22,12 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import carpet.helpers.ContainerAutoCraftingTable;
+import carpet.logging.LoggerRegistry;
+import carpet.utils.Messenger;
+import net.minecraft.item.crafting.CraftingManager;
+import net.minecraft.util.text.ITextComponent;
+
 public class ServerRecipeBookHelper
 {
     private final Logger field_194330_a = LogManager.getLogger();
@@ -35,6 +41,18 @@
 
     public void func_194327_a(EntityPlayerMP p_194327_1_, @Nullable IRecipe p_194327_2_, boolean p_194327_3_)
     {
+
+        if(LoggerRegistry.__recipes){ // Added debugger for the instance people need help debuging why there recipes don't work. CARPET-XCOM
+            int i = 0;
+            for(ResourceLocation r : CraftingManager.field_193380_a.func_148742_b()) {
+                String index = Integer.toString(i);
+                LoggerRegistry.getLogger("recipes").log(()-> new ITextComponent[]{
+                        Messenger.s(null, index + ": " + r.func_110623_a())
+                });
+                i++;
+            }
+        }
+
         if (p_194327_2_ != null && p_194327_1_.func_192037_E().func_193830_f(p_194327_2_))
         {
             this.field_194332_c = p_194327_1_;
@@ -48,7 +66,10 @@
             if (container instanceof ContainerWorkbench)
             {
                 this.field_194335_f = ((ContainerWorkbench)container).field_75160_f;
-                this.field_194336_g = ((ContainerWorkbench)container).field_75162_e;
+                // Method added for auto crafting CARPET-XCOM
+                this.field_194336_g = container instanceof ContainerAutoCraftingTable
+                        ? ((ContainerAutoCraftingTable) container).getInventoryCrafting()
+                        : ((ContainerWorkbench)container).field_75162_e;
             }
             else if (container instanceof ContainerPlayer)
             {
