--- ../src-base/minecraft/net/minecraft/util/Util.java
+++ ../src-work/minecraft/net/minecraft/util/Util.java
@@ -4,6 +4,9 @@
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.FutureTask;
 import javax.annotation.Nullable;
+
+import carpet.CarpetSettings;
+import carpet.helpers.ThrowableSuppression;
 import org.apache.logging.log4j.Logger;
 
 public class Util
@@ -18,7 +21,10 @@
         }
         catch (ExecutionException executionexception)
         {
-            p_181617_1_.fatal("Error executing task", (Throwable)executionexception);
+            // Update Suppression crash fixes CARPET-XCOM
+            if(!CarpetSettings.updateSuppressionCrashFix || !(executionexception.getCause() instanceof ThrowableSuppression))
+                p_181617_1_.fatal("Error executing task", (Throwable)executionexception);
+            net.minecraft.server.management.PlayerInteractionManager.playerMinedBlock = null;
         }
         catch (InterruptedException interruptedexception)
         {
