--- ../src-base/minecraft/net/minecraft/network/NetHandlerPlayServer.java
+++ ../src-work/minecraft/net/minecraft/network/NetHandlerPlayServer.java
@@ -121,6 +121,15 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import carpet.CarpetServer;
+import carpet.CarpetSettings;
+import carpet.helpers.TickSpeed;
+import carpet.worldedit.WorldEditBridge;
+import net.minecraft.world.gen.structure.template.Template;
+import net.minecraft.server.management.PlayerInteractionManager;
+
+import static carpet.commands.CommandGMS.setPlayerToSurvival;
+
 public class NetHandlerPlayServer implements INetHandlerPlayServer, ITickable
 {
     private static final Logger field_147370_c = LogManager.getLogger();
@@ -257,9 +266,19 @@
         {
             this.func_194028_b(new TextComponentTranslation("multiplayer.disconnect.idling", new Object[0]));
         }
+
+        // [CM] PistonClippingFix -- Check if counter is above 0 and decrement
+        if(this.field_147369_b.pistonClippingCounter > 0 && CarpetSettings.pistonClippingFix > 0)
+        {
+            this.field_147369_b.pistonClippingCounter--;
+        }
+        else
+        {
+            this.field_147369_b.pistonClippingCounter = 0;
+        }
     }
 
-    private void func_184342_d()
+    public void func_184342_d()
     {
         this.field_184349_l = this.field_147369_b.field_70165_t;
         this.field_184350_m = this.field_147369_b.field_70163_u;
@@ -297,6 +316,11 @@
     {
         PacketThreadUtil.func_180031_a(p_147358_1_, this, this.field_147369_b.func_71121_q());
         this.field_147369_b.func_110430_a(p_147358_1_.func_149620_c(), p_147358_1_.func_192620_b(), p_147358_1_.func_149618_e(), p_147358_1_.func_149617_f());
+        //CM checking if player is moving, for commandTick
+        if (p_147358_1_.func_149620_c() != 0.0F || p_147358_1_.func_192620_b() != 0.0F || p_147358_1_.func_149618_e() || p_147358_1_.func_149617_f())
+        {
+            TickSpeed.reset_player_active_timeout();
+        }
     }
 
     private static boolean func_183006_b(CPacketPlayer p_183006_0_)
@@ -500,6 +524,10 @@
                         double d10 = this.field_147369_b.field_70159_w * this.field_147369_b.field_70159_w + this.field_147369_b.field_70181_x * this.field_147369_b.field_70181_x + this.field_147369_b.field_70179_y * this.field_147369_b.field_70179_y;
                         double d11 = d7 * d7 + d8 * d8 + d9 * d9;
 
+                        if (d11 > 0.0001D) // for commandTick
+                        {
+                            TickSpeed.reset_player_active_timeout();
+                        }
                         if (this.field_147369_b.func_70608_bn())
                         {
                             if (d11 > 1.0D)
@@ -564,7 +592,9 @@
                             this.field_147369_b.func_70080_a(d4, d5, d6, f, f1);
                             this.field_147369_b.func_71000_j(this.field_147369_b.field_70165_t - d0, this.field_147369_b.field_70163_u - d1, this.field_147369_b.field_70161_v - d2);
 
-                            if (!this.field_147369_b.field_70145_X && !this.field_147369_b.func_70608_bn())
+                            // [CM] PistonClippingFix -- Added PistonClippingcounter check to the if statement
+                            // Added fix to no clipping creative players. Carpet-XCOM
+                            if (!this.field_147369_b.field_70145_X && !this.field_147369_b.func_70608_bn() && this.field_147369_b.pistonClippingCounter == 0 && (!CarpetSettings.creativeNoClip || !field_147369_b.func_184812_l_()))
                             {
                                 boolean flag1 = worldserver.func_184144_a(this.field_147369_b, this.field_147369_b.func_174813_aQ().func_186664_h(0.0625D)).isEmpty();
 
@@ -731,7 +761,10 @@
         {
             if (this.field_184362_y == null && this.field_147369_b.func_70092_e((double)blockpos.func_177958_n() + 0.5D, (double)blockpos.func_177956_o() + 0.5D, (double)blockpos.func_177952_p() + 0.5D) < 64.0D && !this.field_147367_d.func_175579_a(worldserver, blockpos, this.field_147369_b) && worldserver.func_175723_af().func_177746_a(blockpos))
             {
+                if(carpet.carpetclient.CarpetClientChunkLogger.logger.enabled)
+                    carpet.carpetclient.CarpetClientChunkLogger.setReason("Player interacting with right click");
                 this.field_147369_b.field_71134_c.func_187251_a(this.field_147369_b, worldserver, itemstack, enumhand, blockpos, enumfacing, p_184337_1_.func_187026_d(), p_184337_1_.func_187025_e(), p_184337_1_.func_187020_f());
+                carpet.carpetclient.CarpetClientChunkLogger.resetReason();
             }
         }
         else
@@ -796,13 +829,13 @@
                     this.field_147369_b.field_71093_bK = entity.field_71093_bK;
                     this.func_147359_a(new SPacketRespawn(this.field_147369_b.field_71093_bK, worldserver1.func_175659_aa(), worldserver1.func_72912_H().func_76067_t(), this.field_147369_b.field_71134_c.func_73081_b()));
                     this.field_147367_d.func_184103_al().func_187243_f(this.field_147369_b);
-                    worldserver1.func_72973_f(this.field_147369_b);
+                    worldserver1.func_72900_e(this.field_147369_b);
+                    worldserver1.func_72964_e(this.field_147369_b.field_70176_ah, this.field_147369_b.field_70164_aj).func_76608_a(this.field_147369_b, this.field_147369_b.field_70162_ai);
                     this.field_147369_b.field_70128_L = false;
                     this.field_147369_b.func_70012_b(entity.field_70165_t, entity.field_70163_u, entity.field_70161_v, entity.field_70177_z, entity.field_70125_A);
 
                     if (this.field_147369_b.func_70089_S())
                     {
-                        worldserver1.func_72866_a(this.field_147369_b, false);
                         worldserver2.func_72838_d(this.field_147369_b);
                         worldserver2.func_72866_a(this.field_147369_b, false);
                     }
@@ -1019,14 +1052,17 @@
                 break;
             case START_FALL_FLYING:
 
-                if (!this.field_147369_b.field_70122_E && this.field_147369_b.field_70181_x < 0.0D && !this.field_147369_b.func_184613_cA() && !this.field_147369_b.func_70090_H())
+                if (!this.field_147369_b.field_70122_E && !this.field_147369_b.func_184613_cA() && !this.field_147369_b.func_70090_H())
                 {
+                    // Remove the falling check as in 1.15 CARPET-XCOM
+                    if(CarpetSettings.elytraCheckFix || this.field_147369_b.field_70181_x < 0.0D) {
                     ItemStack itemstack = this.field_147369_b.func_184582_a(EntityEquipmentSlot.CHEST);
 
                     if (itemstack.func_77973_b() == Items.field_185160_cR && ItemElytra.func_185069_d(itemstack))
                     {
                         this.field_147369_b.func_184847_M();
                     }
+                    }
                 }
                 else
                 {
@@ -1147,11 +1183,15 @@
             }
             else
             {
+                CarpetServer.playerInventoryStacking = true;
                 ItemStack itemstack2 = this.field_147369_b.field_71070_bA.func_184996_a(p_147351_1_.func_149544_d(), p_147351_1_.func_149543_e(), p_147351_1_.func_186993_f(), this.field_147369_b);
+                CarpetServer.playerInventoryStacking = false;
 
                 if (ItemStack.func_77989_b(p_147351_1_.func_149546_g(), itemstack2))
                 {
                     this.field_147369_b.field_71135_a.func_147359_a(new SPacketConfirmTransaction(p_147351_1_.func_149548_c(), p_147351_1_.func_149547_f(), true));
+                    // Update item changes before setting boolean true given it can cause desynchs. CARPET-XCOM
+                    if(CarpetSettings.itemDesynchFix) this.field_147369_b.field_71070_bA.func_75142_b();
                     this.field_147369_b.field_71137_h = true;
                     this.field_147369_b.field_71070_bA.func_75142_b();
                     this.field_147369_b.func_71113_k();
