--- ../src-base/minecraft/net/minecraft/item/ItemShears.java
+++ ../src-work/minecraft/net/minecraft/item/ItemShears.java
@@ -9,6 +9,10 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 
+import carpet.CarpetSettings;
+
+ import carpet.CarpetSettings;
+
 public class ItemShears extends Item
 {
     public ItemShears()
@@ -38,8 +42,11 @@
     public float func_150893_a(ItemStack p_150893_1_, IBlockState p_150893_2_)
     {
         Block block = p_150893_2_.func_177230_c();
-
-        if (block != Blocks.field_150321_G && p_150893_2_.func_185904_a() != Material.field_151584_j)
+        
+        /*
+         * Added missing tool for sponge CARPET-XCOM
+         */
+        if (block != Blocks.field_150321_G && p_150893_2_.func_185904_a() != Material.field_151584_j && (!CarpetSettings.missingTools || p_150893_2_.func_185904_a() != Material.field_151583_m))
         {
             return block == Blocks.field_150325_L ? 5.0F : super.func_150893_a(p_150893_1_, p_150893_2_);
         }
