--- ../src-base/minecraft/net/minecraft/item/ItemBlockSpecial.java
+++ ../src-work/minecraft/net/minecraft/item/ItemBlockSpecial.java
@@ -16,6 +16,9 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 
+import carpet.CarpetSettings;
+import carpet.helpers.BlockRotator;
+
 public class ItemBlockSpecial extends Item
 {
     private final Block field_150935_a;
@@ -43,10 +46,25 @@
 
         if (!itemstack.func_190926_b() && p_180614_1_.func_175151_a(p_180614_3_, p_180614_5_, itemstack) && p_180614_2_.func_190527_a(this.field_150935_a, p_180614_3_, false, p_180614_5_, (Entity)null))
         {
-            IBlockState iblockstate1 = this.field_150935_a.func_180642_a(p_180614_2_, p_180614_3_, p_180614_5_, p_180614_6_, p_180614_7_, p_180614_8_, 0, p_180614_1_);
 
+            //Carpet Mod Accurate block placement replacement code
+            IBlockState iblockstate1 = null;
+            if(CarpetSettings.accurateBlockPlacement && p_180614_6_ > 1)
+            {
+                iblockstate1 = BlockRotator.alternativeBlockPlacement(this.field_150935_a, p_180614_2_, p_180614_3_, p_180614_5_, p_180614_6_, p_180614_7_, p_180614_8_, 0, p_180614_1_);
+            }
+            if (iblockstate1 == null) // vanilla
+            {
+                iblockstate1 = this.field_150935_a.func_180642_a(p_180614_2_, p_180614_3_, p_180614_5_, p_180614_6_ % 2.0F, p_180614_7_, p_180614_8_, 0, p_180614_1_);
+            }
+            //IBlockState iblockstate1 = this.block.onBlockPlaced(playerIn, worldIn, hand, facing, hitX, hitY, 0, stack); //this got replaced
+            //END CM
+
+
+            if(CarpetSettings.duplicationFixUpdateSuppression) itemstack.func_190918_g(1);
             if (!p_180614_2_.func_180501_a(p_180614_3_, iblockstate1, 11))
             {
+                if(CarpetSettings.duplicationFixUpdateSuppression) itemstack.func_190918_g(-1); // Fix for duping when update suppressing. CARPET-XCOM
                 return EnumActionResult.FAIL;
             }
             else
@@ -66,7 +84,7 @@
 
                 SoundType soundtype = this.field_150935_a.func_185467_w();
                 p_180614_2_.func_184133_a(p_180614_1_, p_180614_3_, soundtype.func_185841_e(), SoundCategory.BLOCKS, (soundtype.func_185843_a() + 1.0F) / 2.0F, soundtype.func_185847_b() * 0.8F);
-                itemstack.func_190918_g(1);
+                if(!CarpetSettings.duplicationFixUpdateSuppression) itemstack.func_190918_g(1);
                 return EnumActionResult.SUCCESS;
             }
         }
