--- ../src-base/minecraft/net/minecraft/item/ItemSign.java
+++ ../src-work/minecraft/net/minecraft/item/ItemSign.java
@@ -1,5 +1,6 @@
 package net.minecraft.item;
 
+import carpet.CarpetSettings;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.block.BlockStandingSign;
 import net.minecraft.block.BlockWallSign;
@@ -43,6 +44,8 @@
                 }
                 else
                 {
+                    if(CarpetSettings.duplicationFixUpdateSuppression) itemstack.func_190918_g(1); // Fix for duping when update suppressing. CARPET-XCOM
+
                     p_180614_3_ = flag ? p_180614_3_.func_177977_b() : p_180614_3_;
 
                     if (p_180614_5_ == EnumFacing.UP)
@@ -67,7 +70,7 @@
                         CriteriaTriggers.field_193137_x.func_193173_a((EntityPlayerMP)p_180614_1_, p_180614_3_, itemstack);
                     }
 
-                    itemstack.func_190918_g(1);
+                    if(!CarpetSettings.duplicationFixUpdateSuppression) itemstack.func_190918_g(1);
                     return EnumActionResult.SUCCESS;
                 }
             }
