--- ../src-base/minecraft/net/minecraft/item/ItemFireball.java
+++ ../src-work/minecraft/net/minecraft/item/ItemFireball.java
@@ -12,6 +12,11 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 
+import carpet.CarpetSettings;
+import carpet.helpers.GhastHelper;
+import net.minecraft.entity.monster.EntityGhast;
+import net.minecraft.util.ActionResult;
+
 public class ItemFireball extends Item
 {
     public ItemFireball()
@@ -51,4 +56,22 @@
             }
         }
     }
+
+
+    public ActionResult<ItemStack> func_77659_a(World itemStackIn, EntityPlayer worldIn, EnumHand playerIn)
+    {
+        if (!(CarpetSettings.rideableGhasts && worldIn.func_184187_bx() instanceof EntityGhast))
+        {
+            return super.func_77659_a(itemStackIn, worldIn, playerIn);
+        }
+        ItemStack itemstack = worldIn.func_184586_b(playerIn);
+        EntityGhast ghast = (EntityGhast)worldIn.func_184187_bx();
+        worldIn.func_184811_cZ().func_185145_a(this, 40);
+        GhastHelper.set_off_fball(ghast, itemStackIn, worldIn);
+        if (!worldIn.field_71075_bZ.field_75098_d)
+        {
+            itemstack.func_190918_g(1);
+        }
+        return new ActionResult<>(EnumActionResult.SUCCESS, itemstack);
+    }
 }
