--- ../src-base/minecraft/net/minecraft/command/CommandClone.java
+++ ../src-work/minecraft/net/minecraft/command/CommandClone.java
@@ -18,6 +18,10 @@
 import net.minecraft.world.World;
 import net.minecraft.world.gen.structure.StructureBoundingBox;
 
+import carpet.CarpetSettings;
+import carpet.worldedit.WorldEditBridge;
+import net.minecraft.entity.player.EntityPlayerMP;
+
 public class CommandClone extends CommandBase
 {
     public String func_71517_b()
@@ -51,9 +55,9 @@
             StructureBoundingBox structureboundingbox1 = new StructureBoundingBox(blockpos2, blockpos2.func_177971_a(structureboundingbox.func_175896_b()));
             int i = structureboundingbox.func_78883_b() * structureboundingbox.func_78882_c() * structureboundingbox.func_78880_d();
 
-            if (i > 32768)
+            if (i > CarpetSettings.fillLimit) //CM
             {
-                throw new CommandException("commands.clone.tooManyBlocks", new Object[] {i, Integer.valueOf(32768)});
+                throw new CommandException("commands.clone.tooManyBlocks", new Object[] {i, Integer.valueOf(CarpetSettings.fillLimit)}); //CM
             }
             else
             {
@@ -143,10 +147,13 @@
                                 }
                             }
 
+                            EntityPlayerMP worldEditPlayer = p_184881_2_ instanceof EntityPlayerMP ? (EntityPlayerMP) p_184881_2_ : null;
+                            
                             if (flag)
                             {
                                 for (BlockPos blockpos6 : deque)
                                 {
+                                    WorldEditBridge.recordBlockEdit(worldEditPlayer, world, blockpos6, Blocks.field_150350_a.func_176223_P(), null);
                                     TileEntity tileentity1 = world.func_175625_s(blockpos6);
 
                                     if (tileentity1 instanceof IInventory)
@@ -154,12 +161,12 @@
                                         ((IInventory)tileentity1).func_174888_l();
                                     }
 
-                                    world.func_180501_a(blockpos6, Blocks.field_180401_cv.func_176223_P(), 2);
+                                    world.func_180501_a(blockpos6, Blocks.field_180401_cv.func_176223_P(), 2 | (CarpetSettings.fillUpdates?0:128)); //carpet
                                 }
 
                                 for (BlockPos blockpos7 : deque)
                                 {
-                                    world.func_180501_a(blockpos7, Blocks.field_150350_a.func_176223_P(), 3);
+                                    world.func_180501_a(blockpos7, Blocks.field_150350_a.func_176223_P(), (CarpetSettings.fillUpdates?3:131)); //carpet
                                 }
                             }
 
@@ -171,6 +178,7 @@
 
                             for (CommandClone.StaticCloneData commandclone$staticclonedata : list4)
                             {
+                                WorldEditBridge.recordBlockEdit(worldEditPlayer, world, commandclone$staticclonedata.field_179537_a, commandclone$staticclonedata.field_179535_b, commandclone$staticclonedata.field_184953_c);
                                 TileEntity tileentity2 = world.func_175625_s(commandclone$staticclonedata.field_179537_a);
 
                                 if (tileentity2 instanceof IInventory)
@@ -178,19 +186,18 @@
                                     ((IInventory)tileentity2).func_174888_l();
                                 }
 
-                                world.func_180501_a(commandclone$staticclonedata.field_179537_a, Blocks.field_180401_cv.func_176223_P(), 2);
+                                world.func_180501_a(commandclone$staticclonedata.field_179537_a, Blocks.field_180401_cv.func_176223_P(), 2 | (CarpetSettings.fillUpdates?0:128)); //carpet
                             }
 
                             i = 0;
 
                             for (CommandClone.StaticCloneData commandclone$staticclonedata1 : list3)
                             {
-                                if (world.func_180501_a(commandclone$staticclonedata1.field_179537_a, commandclone$staticclonedata1.field_179535_b, 2))
+                                if (world.func_180501_a(commandclone$staticclonedata1.field_179537_a, commandclone$staticclonedata1.field_179535_b, 2 | (CarpetSettings.fillUpdates?0:128))) //carpet
                                 {
                                     ++i;
                                 }
                             }
-
                             for (CommandClone.StaticCloneData commandclone$staticclonedata2 : list1)
                             {
                                 TileEntity tileentity3 = world.func_175625_s(commandclone$staticclonedata2.field_179537_a);
@@ -206,7 +213,11 @@
 
                                 world.func_180501_a(commandclone$staticclonedata2.field_179537_a, commandclone$staticclonedata2.field_179535_b, 2);
                             }
-
+                            
+                            /*carpet mod */
+                            if (CarpetSettings.fillUpdates)
+                            {
+                                /*carpet mod end EXTRA INDENTATION START*/
                             for (CommandClone.StaticCloneData commandclone$staticclonedata3 : list4)
                             {
                                 world.func_175722_b(commandclone$staticclonedata3.field_179537_a, commandclone$staticclonedata3.field_179535_b.func_177230_c(), false);
@@ -225,7 +236,8 @@
                                     }
                                 }
                             }
-
+                            } //carpet mod back extra indentation
+                            
                             if (i <= 0)
                             {
                                 throw new CommandException("commands.clone.failed", new Object[0]);
