--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenLiquids.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenLiquids.java
@@ -1,11 +1,17 @@
 package net.minecraft.world.gen.feature;
 
 import java.util.Random;
+
+import carpet.CarpetSettings;
+import carpet.logging.LoggerRegistry;
+import carpet.utils.Messenger;
 import net.minecraft.block.Block;
 import net.minecraft.block.material.Material;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.init.Blocks;
 import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.text.ITextComponent;
+import net.minecraft.util.text.TextFormatting;
 import net.minecraft.world.World;
 
 public class WorldGenLiquids extends WorldGenerator
@@ -19,6 +25,13 @@
 
     public boolean func_180709_b(World p_180709_1_, Random p_180709_2_, BlockPos p_180709_3_)
     {
+        if (LoggerRegistry.__liquidPlacement && p_180709_3_.func_177956_o() >= CarpetSettings.logLiquidPlacementMinHeight && p_180709_3_.func_177956_o() <= CarpetSettings.logLiquidPlacementMaxHeight) {
+            TextFormatting color = field_150521_a.equals(Blocks.field_150356_k) ? TextFormatting.BLUE: TextFormatting.RED;;
+            LoggerRegistry.getLogger("liquidPlacement").logNoCommand(()-> new ITextComponent[]{
+                    Messenger.s(null, TextFormatting.GREEN + "- Pocket of " + color + field_150521_a + TextFormatting.GREEN + " tried to generate in: " +
+                            TextFormatting.AQUA + p_180709_3_)
+            });
+        }
         if (p_180709_1_.func_180495_p(p_180709_3_.func_177984_a()).func_177230_c() != Blocks.field_150348_b)
         {
             return false;
